<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="utf-8">
    <link href="https://fonts.googleapis.com/icon?family=Material+Icons" rel="stylesheet"> 
    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.6.3/css/all.css" integrity="sha384-UHRtZLI+pbxtHCWp1t77Bi1L4ZtiqrqD80Kn4Z8NTSRyMA2Fd33n5dQ8lWUE00s/" crossorigin="anonymous">
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css"> 
    
    <link rel="stylesheet" href="../../../fonts/academicons-1.8.6/css/academicons.min.css"/>
    <link rel="icon" type="image/png" sizes="32x32" href="../../../logo/bodhi.png"> 
    <meta name="viewport" content="width=device-width, initial-scale=1">
    
    
    
    <title>UCSCXenaTools: Retrieve Gene Expression and Clinical Information from UCSC Xena for Survival Analysis - Shixiang Wang</title>
    
     
    <meta property="og:title" content="UCSCXenaTools: Retrieve Gene Expression and Clinical Information from UCSC Xena for Survival Analysis - Shixiang Wang | 王诗翔">
    

    
      
    

    

    
    


<link href='//cdn.bootcss.com/highlight.js/9.12.0/styles/github.min.css' rel='stylesheet' type='text/css' />



    <link rel="stylesheet" href="../../../css/style.css" />
    <link rel="stylesheet" href="../../../css/mystyle.css" /> 
    <link rel="stylesheet" href="../../../css/fonts.css" />
    
<link rel="stylesheet" href="../../../css/custom.css" />

  </head>

  
  <body class="en">
    <header class="masthead">
      

<h1><a href="../../../"><img src="https://avatars1.githubusercontent.com/u/25057508?s=460&amp;v=4" alt="Shixiang Wang" /></a></h1>



      <nav class="menu">
        <input id="menu-check" type="checkbox" />
        <label id="menu-label" for="menu-check" class="unselectable">
          <span class="icon close-icon">✕</span>
          <span class="icon open-icon">☰</span>
          <span class="text">Menu</span>
        </label>
        <ul>
        
        
        <li><a href="../../../">Home</a></li>
        
        <li><a href="../../../en/about/">About</a></li>
        
        <li><a href="../../../en/post/">Blog</a></li>
        
        <li><a href="../../../en/tools/">Software</a></li>
        
        <li><a href="../../../en/list/">List</a></li>
        
        <li><a href="../../../en/cv-en/shixiang">CV</a></li>
        
        <li><a href="../../../cn/">中文</a></li>
        
        

<li class="menu-extra"></li>






<li><a href="https://github.com/ShixiangWang/home/edit/master/content/en/post/2019-08-20-UCSCXenaTools-notes-to-rOpenSci.Rmd" target="_blank">Edit Me</a></li>


<li><a href="../../../en/index.xml" type="application/rss+xml" title="RSS feed">Subscribe</a></li>

<li><a href="http://creativecommons.org/licenses/by-nc-sa/4.0/" title="Attribution-NonCommercial-ShareAlike 4.0 International">License</a></li>


        </ul>
      </nav>
    </header>

    <article class="main">
      <header class="title">
        

<h1>UCSCXenaTools: Retrieve Gene Expression and Clinical Information from UCSC Xena for Survival Analysis</h1>



<h3>Shixiang Wang &middot 
2019-08-25</h3> 


   
  


      </header>





<pre class="r"><code>knitr::opts_chunk$set(comment = &quot;#&gt;&quot;, collapse = TRUE)</code></pre>
<p>In this technote I will outline how to use <strong>UCSCXenaTools</strong> package to pull gene expression and clinical data from UCSC Xena for survival analysis.</p>
<p>For general usage of <strong>UCSCXenaTools</strong>, please refer to <a href="https://cran.r-project.org/web/packages/UCSCXenaTools/vignettes/USCSXenaTools.html">package vignette</a>.
Any bug or feature request can be reported in <a href="https://github.com/ropensci/UCSCXenaTools/issues">GitHub issues</a>.</p>
<div id="installation" class="section level2">
<h2>Installation</h2>
<p><strong>UCSCXenaTools</strong> is available from CRAN:</p>
<pre class="r"><code>install.packages(&quot;UCSCXenaTools&quot;)</code></pre>
<p>Alternatively, the latest development version can be downloaded from GitHub:</p>
<pre class="r"><code>remotes::install_github(&quot;ropensci/UCSCXenaTools&quot;, build_vignettes = TRUE, dependencies = TRUE)</code></pre>
</div>
<div id="how-it-works" class="section level2">
<h2>How it works</h2>
<p>Before actually pulling data, understand how <strong>UCSCXenaTools</strong> works (see Figure 1) will help user locate the most important function to use.</p>
<p>Generally,</p>
<ul>
<li>for operating datasets, we use functions whose names start with <code>Xena</code></li>
<li>for operating subset of a dataset, we use functions whose names start with <code>fetch_</code></li>
</ul>
<p><img src="https://raw.githubusercontent.com/ropensci/UCSCXenaTools/master/paper/overview.png" />
<em>Figure 1. The UCSCXenaTools pipeline</em></p>
</div>
<div id="download-data" class="section level2">
<h2>Download data</h2>
<p>To illustrate how to download data and do survival analysis, here we use
expression data of gene <a href="https://ghr.nlm.nih.gov/gene/TP53"><em>TP53</em></a> (which is a famous tumor suppressor) and survival status from <a href="https://xenabrowser.net/datapages/?cohort=TCGA%20Lung%20Adenocarcinoma%20(LUAD)&amp;removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443">TCGA Lung Adenocarcinoma (LUAD)</a> as example data.</p>
<p>For obtaining survival status of TCGA LUAD samples, we need to download the TCGA LUAD clinical dataset.
For obtaining <em>TP53</em> gene expression, we need to download a subset of TCGA LUAD expression matrix.</p>
<pre class="r"><code>suppressMessages(library(UCSCXenaTools))
suppressMessages(library(dplyr))</code></pre>
<pre class="r"><code>luad_cohort = XenaData %&gt;% 
  filter(XenaHostNames == &quot;tcgaHub&quot;) %&gt;% # select TCGA Hub
  XenaScan(&quot;TCGA Lung Adenocarcinoma&quot;)   # select LUAD cohort

luad_cohort
#&gt; # A tibble: 27 x 17
#&gt;    XenaHosts XenaHostNames XenaCohorts XenaDatasets SampleCount DataSubtype
#&gt;    &lt;chr&gt;     &lt;chr&gt;         &lt;chr&gt;       &lt;chr&gt;        &lt;chr&gt;       &lt;chr&gt;      
#&gt;  1 https://… tcgaHub       TCGA Lung … RABIT/separ… 467         Transcript…
#&gt;  2 https://… tcgaHub       TCGA Lung … RABIT/separ… 120         Transcript…
#&gt;  3 https://… tcgaHub       TCGA Lung … TCGA.LUAD.s… 151         DNA methyl…
#&gt;  4 https://… tcgaHub       TCGA Lung … TCGA.LUAD.s… 492         DNA methyl…
#&gt;  5 https://… tcgaHub       TCGA Lung … TCGA.LUAD.s… 516         copy numbe…
#&gt;  6 https://… tcgaHub       TCGA Lung … TCGA.LUAD.s… 543         somatic mu…
#&gt;  7 https://… tcgaHub       TCGA Lung … TCGA.LUAD.s… 237         protein ex…
#&gt;  8 https://… tcgaHub       TCGA Lung … TCGA.LUAD.s… 576         gene expre…
#&gt;  9 https://… tcgaHub       TCGA Lung … TCGA.LUAD.s… 60          miRNA matu…
#&gt; 10 https://… tcgaHub       TCGA Lung … TCGA.LUAD.s… 576         gene expre…
#&gt; # … with 17 more rows, and 11 more variables: Label &lt;chr&gt;, Type &lt;chr&gt;,
#&gt; #   AnatomicalOrigin &lt;chr&gt;, SampleType &lt;chr&gt;, Tags &lt;chr&gt;, ProbeMap &lt;chr&gt;,
#&gt; #   LongTitle &lt;chr&gt;, Citation &lt;chr&gt;, Version &lt;chr&gt;, Unit &lt;chr&gt;,
#&gt; #   Platform &lt;chr&gt;</code></pre>
<p><code>luad_cohort</code> contains information of all datasets in TCGA LUAD cohort.</p>
<div id="download-clinical-dataset" class="section level3">
<h3>Download clinical dataset</h3>
<p>Now we download clinical dataset of TCGA LUAD cohort and load it into R.</p>
<pre class="r"><code>cli_query = luad_cohort %&gt;% 
  filter(DataSubtype == &quot;phenotype&quot;) %&gt;%  # select clinical dataset
  XenaGenerate() %&gt;%  # generate a XenaHub object
  XenaQuery() %&gt;% 
  XenaDownload()
#&gt; This will check url status, please be patient.
#&gt; All downloaded files will under directory /Volumes/Data/Rtmp/RtmpairAff.
#&gt; The &#39;trans_slash&#39; option is FALSE, keep same directory structure as Xena.
#&gt; Creating directories for datasets...
#&gt; Downloading TCGA.LUAD.sampleMap/LUAD_clinicalMatrix.gz

cli = XenaPrepare(cli_query)

# See a few rows
head(cli)
#&gt; # A tibble: 6 x 157
#&gt;   sampleID ABSOLUTE_Ploidy ABSOLUTE_Purity AKT1  ALK_translocati… BRAF 
#&gt;   &lt;chr&gt;              &lt;dbl&gt;           &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;            &lt;chr&gt;
#&gt; 1 TCGA-05…           NA              NA    &lt;NA&gt;  &lt;NA&gt;             &lt;NA&gt; 
#&gt; 2 TCGA-05…            3.77            0.46 none  &lt;NA&gt;             p.A7…
#&gt; 3 TCGA-05…           NA              NA    &lt;NA&gt;  &lt;NA&gt;             &lt;NA&gt; 
#&gt; 4 TCGA-05…           NA              NA    none  &lt;NA&gt;             p.L6…
#&gt; 5 TCGA-05…            2.04            0.48 none  &lt;NA&gt;             none 
#&gt; 6 TCGA-05…            3.29            0.48 none  &lt;NA&gt;             p.G4…
#&gt; # … with 151 more variables: CBL &lt;chr&gt;, CTNNB1 &lt;chr&gt;,
#&gt; #   Canonical_mut_in_KRAS_EGFR_ALK &lt;chr&gt;,
#&gt; #   Cnncl_mt_n_KRAS_EGFR_ALK_RET_ROS1_BRAF_ERBB2_HRAS_NRAS_AKT1_MAP2 &lt;chr&gt;,
#&gt; #   EGFR &lt;chr&gt;, ERBB2 &lt;chr&gt;, ERBB4 &lt;chr&gt;,
#&gt; #   Estimated_allele_fraction_of_a_clonal_varnt_prsnt_t_1_cpy_pr_cll &lt;dbl&gt;,
#&gt; #   Expression_Subtype &lt;chr&gt;, HRAS &lt;chr&gt;, KRAS &lt;chr&gt;, MAP2K1 &lt;chr&gt;,
#&gt; #   MET &lt;chr&gt;, NRAS &lt;chr&gt;, PIK3CA &lt;chr&gt;, PTPN11 &lt;chr&gt;, Pathology &lt;chr&gt;,
#&gt; #   Pathology_Updated &lt;chr&gt;, RET_translocation &lt;chr&gt;,
#&gt; #   ROS1_translocation &lt;chr&gt;, STK11 &lt;chr&gt;,
#&gt; #   WGS_as_of_20120731_0_no_1_yes &lt;dbl&gt;, `_EVENT` &lt;dbl&gt;,
#&gt; #   `_INTEGRATION` &lt;chr&gt;, OS.time &lt;dbl&gt;, OS &lt;dbl&gt;, OS.unit &lt;chr&gt;,
#&gt; #   `_PANCAN_CNA_PANCAN_K8` &lt;chr&gt;, `_PANCAN_Cluster_Cluster_PANCAN` &lt;chr&gt;,
#&gt; #   `_PANCAN_DNAMethyl_LUAD` &lt;chr&gt;, `_PANCAN_DNAMethyl_PANCAN` &lt;chr&gt;,
#&gt; #   `_PANCAN_RPPA_PANCAN_K8` &lt;chr&gt;, `_PANCAN_UNC_RNAseq_PANCAN_K16` &lt;chr&gt;,
#&gt; #   `_PANCAN_miRNA_PANCAN` &lt;chr&gt;, `_PANCAN_mirna_LUAD` &lt;chr&gt;,
#&gt; #   `_PANCAN_mutation_PANCAN` &lt;chr&gt;, `_PATIENT` &lt;chr&gt;, RFS.time &lt;dbl&gt;,
#&gt; #   RFS &lt;dbl&gt;, RFS.unit &lt;chr&gt;, `_TIME_TO_EVENT` &lt;dbl&gt;,
#&gt; #   `_TIME_TO_EVENT_UNIT` &lt;chr&gt;, `_cohort` &lt;chr&gt;,
#&gt; #   `_primary_disease` &lt;chr&gt;, `_primary_site` &lt;chr&gt;,
#&gt; #   additional_pharmaceutical_therapy &lt;chr&gt;,
#&gt; #   additional_radiation_therapy &lt;chr&gt;,
#&gt; #   additional_surgery_locoregional_procedure &lt;chr&gt;,
#&gt; #   additional_surgery_metastatic_procedure &lt;chr&gt;,
#&gt; #   age_at_initial_pathologic_diagnosis &lt;dbl&gt;,
#&gt; #   anatomic_neoplasm_subdivision &lt;chr&gt;,
#&gt; #   anatomic_neoplasm_subdivision_other &lt;chr&gt;, bcr_followup_barcode &lt;chr&gt;,
#&gt; #   bcr_patient_barcode &lt;chr&gt;, bcr_sample_barcode &lt;chr&gt;,
#&gt; #   days_to_additional_surgery_locoregional_procedure &lt;dbl&gt;,
#&gt; #   days_to_additional_surgery_metastatic_procedure &lt;dbl&gt;,
#&gt; #   days_to_birth &lt;dbl&gt;, days_to_collection &lt;dbl&gt;, days_to_death &lt;dbl&gt;,
#&gt; #   days_to_initial_pathologic_diagnosis &lt;dbl&gt;,
#&gt; #   days_to_last_followup &lt;dbl&gt;,
#&gt; #   days_to_new_tumor_event_after_initial_treatment &lt;dbl&gt;,
#&gt; #   disease_code &lt;chr&gt;, dlco_predictive_percent &lt;dbl&gt;,
#&gt; #   eastern_cancer_oncology_group &lt;dbl&gt;, egfr_mutation_performed &lt;chr&gt;,
#&gt; #   egfr_mutation_result &lt;chr&gt;, eml4_alk_translocation_method &lt;chr&gt;,
#&gt; #   eml4_alk_translocation_performed &lt;chr&gt;,
#&gt; #   followup_case_report_form_submission_reason &lt;chr&gt;,
#&gt; #   followup_treatment_success &lt;chr&gt;, form_completion_date &lt;chr&gt;,
#&gt; #   gender &lt;chr&gt;, histological_type &lt;chr&gt;,
#&gt; #   history_of_neoadjuvant_treatment &lt;chr&gt;, icd_10 &lt;chr&gt;,
#&gt; #   icd_o_3_histology &lt;chr&gt;, icd_o_3_site &lt;chr&gt;,
#&gt; #   informed_consent_verified &lt;chr&gt;, initial_weight &lt;dbl&gt;,
#&gt; #   intermediate_dimension &lt;dbl&gt;, is_ffpe &lt;chr&gt;,
#&gt; #   karnofsky_performance_score &lt;dbl&gt;, kras_gene_analysis_performed &lt;chr&gt;,
#&gt; #   kras_mutation_found &lt;chr&gt;, kras_mutation_result &lt;chr&gt;,
#&gt; #   location_in_lung_parenchyma &lt;chr&gt;, longest_dimension &lt;dbl&gt;,
#&gt; #   lost_follow_up &lt;chr&gt;, new_neoplasm_event_type &lt;chr&gt;,
#&gt; #   new_tumor_event_after_initial_treatment &lt;chr&gt;,
#&gt; #   number_pack_years_smoked &lt;dbl&gt;, oct_embedded &lt;lgl&gt;, other_dx &lt;chr&gt;,
#&gt; #   pathologic_M &lt;chr&gt;, pathologic_N &lt;chr&gt;, pathologic_T &lt;chr&gt;,
#&gt; #   pathologic_stage &lt;chr&gt;, pathology_report_file_name &lt;chr&gt;, …</code></pre>
</div>
<div id="download-tp53-gene-expression" class="section level3">
<h3>Download <em>TP53</em> gene expression</h3>
<p>To download gene expression data, we need select the right dataset firstly.</p>
<pre class="r"><code>ge = luad_cohort %&gt;% 
  filter(DataSubtype == &quot;gene expression RNAseq&quot;, Label == &quot;IlluminaHiSeq&quot;)
ge
#&gt; # A tibble: 1 x 17
#&gt;   XenaHosts XenaHostNames XenaCohorts XenaDatasets SampleCount DataSubtype
#&gt;   &lt;chr&gt;     &lt;chr&gt;         &lt;chr&gt;       &lt;chr&gt;        &lt;chr&gt;       &lt;chr&gt;      
#&gt; 1 https://… tcgaHub       TCGA Lung … TCGA.LUAD.s… 576         gene expre…
#&gt; # … with 11 more variables: Label &lt;chr&gt;, Type &lt;chr&gt;,
#&gt; #   AnatomicalOrigin &lt;chr&gt;, SampleType &lt;chr&gt;, Tags &lt;chr&gt;, ProbeMap &lt;chr&gt;,
#&gt; #   LongTitle &lt;chr&gt;, Citation &lt;chr&gt;, Version &lt;chr&gt;, Unit &lt;chr&gt;,
#&gt; #   Platform &lt;chr&gt;</code></pre>
<p>Now we fetch <em>TP53</em> gene expression values.</p>
<pre class="r"><code>tp53 = fetch_dense_values(host = ge$XenaHosts, 
                          dataset = ge$XenaDatasets, 
                          identifiers = &quot;TP53&quot;,
                          use_probeMap = TRUE) %&gt;% 
  .[1, ]
#&gt; -&gt; Checking identifiers...
#&gt; -&gt; use_probeMap is TRUE, skipping checking identifiers...
#&gt; -&gt; Done.
#&gt; -&gt; Checking samples...
#&gt; -&gt; Done.
#&gt; -&gt; Checking if the dataset has probeMap...
#&gt; -&gt; Done. ProbeMap is found.

head(tp53)
#&gt; TCGA-69-7978-01 TCGA-62-8399-01 TCGA-78-7539-01 TCGA-50-5931-11 
#&gt;           9.895           8.309          10.350           9.620 
#&gt; TCGA-73-4658-01 TCGA-44-6775-01 
#&gt;          10.020          10.160</code></pre>
</div>
<div id="merge-expression-data-and-survival-status" class="section level3">
<h3>Merge expression data and survival status</h3>
<pre class="r"><code>merged_data = tibble(sampleID = names(tp53),
                     tp53_expression = as.numeric(tp53)) %&gt;% 
  left_join(cli, by = &quot;sampleID&quot;) %&gt;% 
  filter(sample_type == &quot;Primary Tumor&quot;)
  

head(merged_data)
#&gt; # A tibble: 6 x 158
#&gt;   sampleID tp53_expression ABSOLUTE_Ploidy ABSOLUTE_Purity AKT1 
#&gt;   &lt;chr&gt;              &lt;dbl&gt;           &lt;dbl&gt;           &lt;dbl&gt; &lt;chr&gt;
#&gt; 1 TCGA-69…            9.90           NA              NA    &lt;NA&gt; 
#&gt; 2 TCGA-62…            8.31           NA              NA    &lt;NA&gt; 
#&gt; 3 TCGA-78…           10.4             2.1             0.69 none 
#&gt; 4 TCGA-73…           10.0             2.01            0.32 none 
#&gt; 5 TCGA-44…           10.2             2.24            0.38 none 
#&gt; 6 TCGA-44…           11.4             2.05            0.3  none 
#&gt; # … with 153 more variables: ALK_translocation &lt;chr&gt;, BRAF &lt;chr&gt;,
#&gt; #   CBL &lt;chr&gt;, CTNNB1 &lt;chr&gt;, Canonical_mut_in_KRAS_EGFR_ALK &lt;chr&gt;,
#&gt; #   Cnncl_mt_n_KRAS_EGFR_ALK_RET_ROS1_BRAF_ERBB2_HRAS_NRAS_AKT1_MAP2 &lt;chr&gt;,
#&gt; #   EGFR &lt;chr&gt;, ERBB2 &lt;chr&gt;, ERBB4 &lt;chr&gt;,
#&gt; #   Estimated_allele_fraction_of_a_clonal_varnt_prsnt_t_1_cpy_pr_cll &lt;dbl&gt;,
#&gt; #   Expression_Subtype &lt;chr&gt;, HRAS &lt;chr&gt;, KRAS &lt;chr&gt;, MAP2K1 &lt;chr&gt;,
#&gt; #   MET &lt;chr&gt;, NRAS &lt;chr&gt;, PIK3CA &lt;chr&gt;, PTPN11 &lt;chr&gt;, Pathology &lt;chr&gt;,
#&gt; #   Pathology_Updated &lt;chr&gt;, RET_translocation &lt;chr&gt;,
#&gt; #   ROS1_translocation &lt;chr&gt;, STK11 &lt;chr&gt;,
#&gt; #   WGS_as_of_20120731_0_no_1_yes &lt;dbl&gt;, `_EVENT` &lt;dbl&gt;,
#&gt; #   `_INTEGRATION` &lt;chr&gt;, OS.time &lt;dbl&gt;, OS &lt;dbl&gt;, OS.unit &lt;chr&gt;,
#&gt; #   `_PANCAN_CNA_PANCAN_K8` &lt;chr&gt;, `_PANCAN_Cluster_Cluster_PANCAN` &lt;chr&gt;,
#&gt; #   `_PANCAN_DNAMethyl_LUAD` &lt;chr&gt;, `_PANCAN_DNAMethyl_PANCAN` &lt;chr&gt;,
#&gt; #   `_PANCAN_RPPA_PANCAN_K8` &lt;chr&gt;, `_PANCAN_UNC_RNAseq_PANCAN_K16` &lt;chr&gt;,
#&gt; #   `_PANCAN_miRNA_PANCAN` &lt;chr&gt;, `_PANCAN_mirna_LUAD` &lt;chr&gt;,
#&gt; #   `_PANCAN_mutation_PANCAN` &lt;chr&gt;, `_PATIENT` &lt;chr&gt;, RFS.time &lt;dbl&gt;,
#&gt; #   RFS &lt;dbl&gt;, RFS.unit &lt;chr&gt;, `_TIME_TO_EVENT` &lt;dbl&gt;,
#&gt; #   `_TIME_TO_EVENT_UNIT` &lt;chr&gt;, `_cohort` &lt;chr&gt;,
#&gt; #   `_primary_disease` &lt;chr&gt;, `_primary_site` &lt;chr&gt;,
#&gt; #   additional_pharmaceutical_therapy &lt;chr&gt;,
#&gt; #   additional_radiation_therapy &lt;chr&gt;,
#&gt; #   additional_surgery_locoregional_procedure &lt;chr&gt;,
#&gt; #   additional_surgery_metastatic_procedure &lt;chr&gt;,
#&gt; #   age_at_initial_pathologic_diagnosis &lt;dbl&gt;,
#&gt; #   anatomic_neoplasm_subdivision &lt;chr&gt;,
#&gt; #   anatomic_neoplasm_subdivision_other &lt;chr&gt;, bcr_followup_barcode &lt;chr&gt;,
#&gt; #   bcr_patient_barcode &lt;chr&gt;, bcr_sample_barcode &lt;chr&gt;,
#&gt; #   days_to_additional_surgery_locoregional_procedure &lt;dbl&gt;,
#&gt; #   days_to_additional_surgery_metastatic_procedure &lt;dbl&gt;,
#&gt; #   days_to_birth &lt;dbl&gt;, days_to_collection &lt;dbl&gt;, days_to_death &lt;dbl&gt;,
#&gt; #   days_to_initial_pathologic_diagnosis &lt;dbl&gt;,
#&gt; #   days_to_last_followup &lt;dbl&gt;,
#&gt; #   days_to_new_tumor_event_after_initial_treatment &lt;dbl&gt;,
#&gt; #   disease_code &lt;chr&gt;, dlco_predictive_percent &lt;dbl&gt;,
#&gt; #   eastern_cancer_oncology_group &lt;dbl&gt;, egfr_mutation_performed &lt;chr&gt;,
#&gt; #   egfr_mutation_result &lt;chr&gt;, eml4_alk_translocation_method &lt;chr&gt;,
#&gt; #   eml4_alk_translocation_performed &lt;chr&gt;,
#&gt; #   followup_case_report_form_submission_reason &lt;chr&gt;,
#&gt; #   followup_treatment_success &lt;chr&gt;, form_completion_date &lt;chr&gt;,
#&gt; #   gender &lt;chr&gt;, histological_type &lt;chr&gt;,
#&gt; #   history_of_neoadjuvant_treatment &lt;chr&gt;, icd_10 &lt;chr&gt;,
#&gt; #   icd_o_3_histology &lt;chr&gt;, icd_o_3_site &lt;chr&gt;,
#&gt; #   informed_consent_verified &lt;chr&gt;, initial_weight &lt;dbl&gt;,
#&gt; #   intermediate_dimension &lt;dbl&gt;, is_ffpe &lt;chr&gt;,
#&gt; #   karnofsky_performance_score &lt;dbl&gt;, kras_gene_analysis_performed &lt;chr&gt;,
#&gt; #   kras_mutation_found &lt;chr&gt;, kras_mutation_result &lt;chr&gt;,
#&gt; #   location_in_lung_parenchyma &lt;chr&gt;, longest_dimension &lt;dbl&gt;,
#&gt; #   lost_follow_up &lt;chr&gt;, new_neoplasm_event_type &lt;chr&gt;,
#&gt; #   new_tumor_event_after_initial_treatment &lt;chr&gt;,
#&gt; #   number_pack_years_smoked &lt;dbl&gt;, oct_embedded &lt;lgl&gt;, other_dx &lt;chr&gt;,
#&gt; #   pathologic_M &lt;chr&gt;, pathologic_N &lt;chr&gt;, pathologic_T &lt;chr&gt;, …</code></pre>
</div>
</div>
<div id="survival-analysis" class="section level2">
<h2>Survival analysis</h2>
</div>


  <footer>
  
<nav class="post-nav">
  <span class="nav-prev">&larr; <a href="../../../en/post/2019-06-20-how-i-create-ucscxenatools-logo/">Create logo for R package UCSCXenaTools with hexSticker</a></span>
  <span class="nav-next"></span>
</nav>
<script type="text/javascript">
document.addEventListener('keyup', function(e) {
  if (e.target.nodeName.toUpperCase() != 'BODY') return;
  var url = false;
  if (e.which == 37) {  
    
    url = '\/en\/post\/2019-06-20-how-i-create-ucscxenatools-logo\/';
    
  } else if (e.which == 39) {  
    
  }
  if (url) window.location = url;
});
</script>



<section class="comments">
  <div id="disqus_thread"></div>
  <script src="../../../js/disqusloader.min.js"></script>
  <script>
  var disqus_config = function () {
  
    this.page.url = "https:\/\/shixiangwang.github.io\/home\/" + location.pathname;
  
  };
  (function() {
    var inIFrame = function() {
      var iframe = true;
      try { iframe = window.self !== window.top; } catch (e) {}
      return iframe;
    };
    if (inIFrame()) return;
    var disqus_js = '//home-xjdzylqrzp.disqus.com/embed.js';
    
    if (location.hash.match(/^#comment/)) {
      var d = document, s = d.createElement('script');
      s.src = disqus_js; s.async = true;
      s.setAttribute('data-timestamp', +new Date());
      (d.head || d.body).appendChild(s);
    } else {
      disqusLoader('#disqus_thread', {
        scriptUrl: disqus_js, laziness: 0, disqusConfig: disqus_config
      });
    }
  })();
  </script>
  <noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
</section>




<script async src="../../../js/fix-toc.js"></script>

<script async src="../../../js/center-img.js"></script>

<script async src="../../../js/right-quote.js"></script>

<script async src="../../../js/no-highlight.js"></script>

<script async src="../../../js/fix-footnote.js"></script>

<script async src="../../../js/math-code.js"></script>

<script async src="../../../js/external-link.js"></script>

<script async src="../../../js/alt-title.js"></script>

<script async src="../../../js/header-link.js"></script>


<script async src="//cdn.bootcss.com/mathjax/2.7.5/MathJax.js?config=TeX-MML-AM_CHTML"></script>

  



<script src="//cdn.bootcss.com/highlight.js/9.12.0/highlight.min.js"></script>



<script src="//cdn.bootcss.com/highlight.js/9.12.0/languages/r.min.js"></script>
<script src="//cdn.bootcss.com/highlight.js/9.12.0/languages/yaml.min.js"></script>
<script src="//cdn.bootcss.com/highlight.js/9.12.0/languages/tex.min.js"></script>
<script>hljs.configure({languages: []}); hljs.initHighlightingOnLoad();</script>




  
  
  

  <div class="copyright"><a href="mailto:w_shixiang@163.com"><i class='far fa-envelope fa-1x'></i></a> &middot; <a href="https://github.com/ShixiangWang"><i class='fab fa-github fa-1x'></i></a> &middot; <a href="https://stackoverflow.com/users/7662327/shixiang-wang"><i class='fab fa-stack-overflow fa-1x'></i></a> &middot; <a href="https://scholar.google.com/citations?user=FvNp0NkAAAAJ&amp;hl=en"><i class='ai ai-google-scholar ai-1x'></i></a> &middot; <a href="https://orcid.org/0000-0002-9275-3557"><i class='ai ai-orcid ai-1x'></i></a> &middot; <a href="https://www.researchgate.net/profile/Wang_Shixiang4"><i class='ai ai-researchgate ai-1x'></i></a> <br> Powered by <a href="https://gohugo.io">Hugo</a> and <a href="https://bookdown.org/yihui/blogdown/">Blogdown</a> &copy; <a href="../../../">Shixiang Wang</a> 2017 - 2019 </div>
  
  

  <script type="text/javascript" src="//rf.revolvermaps.com/0/0/6.js?i=56h9es09xn7&amp;m=6&amp;c=ff0000&amp;cr1=ffffff&amp;f=arial&amp;l=1" async="async"></script>
  </footer>
  </article>
  
  </body>
</html>

