<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="utf-8">
    <link href="https://fonts.googleapis.com/icon?family=Material+Icons" rel="stylesheet"> 
    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.6.3/css/all.css" integrity="sha384-UHRtZLI+pbxtHCWp1t77Bi1L4ZtiqrqD80Kn4Z8NTSRyMA2Fd33n5dQ8lWUE00s/" crossorigin="anonymous">
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css"> 
    
    <link rel="stylesheet" href="../../../fonts/academicons-1.8.6/css/academicons.min.css"/>
    <link rel="icon" type="image/png" sizes="32x32" href="../../../logo/bodhi.png"> 
    <meta name="viewport" content="width=device-width, initial-scale=1">
    
    
    
    <title>Bioconductor分析基因芯片数据 - 王诗翔</title>
    
     
    <meta property="og:title" content="Bioconductor分析基因芯片数据 - Shixiang Wang | 王诗翔">
    

    
      
    

    

    
    


<link href='//cdn.bootcss.com/highlight.js/9.12.0/styles/github.min.css' rel='stylesheet' type='text/css' />



    <link rel="stylesheet" href="../../../css/style.css" />
    <link rel="stylesheet" href="../../../css/mystyle.css" /> 
    <link rel="stylesheet" href="../../../css/fonts.css" />
    
<script async src="../../../js/load-typekit.js"></script>


<link rel="stylesheet" href="../../../css/custom.css" />

  </head>

  
  <body class="cn">
    <header class="masthead">
      

<h1><a href="../../../"><img src="https://avatars1.githubusercontent.com/u/25057508?s=460&amp;v=4" alt="Shixiang Wang" /></a></h1>



      <nav class="menu">
        <input id="menu-check" type="checkbox" />
        <label id="menu-label" for="menu-check" class="unselectable">
          <span class="icon close-icon">✕</span>
          <span class="icon open-icon">☰</span>
          <span class="text">Menu</span>
        </label>
        <ul>
        
        
        <li><a href="../../../">首页</a></li>
        
        <li><a href="../../../cn/about/">关于</a></li>
        
        <li><a href="../../../cn/post/">博客</a></li>
        
        <li><a href="../../../cn/read/">读书</a></li>
        
        <li><a href="../../../cn/writing">写作</a></li>
        
        <li><a href="../../../cn/research/">研究</a></li>
        
        <li><a href="../../../cn/tools/">工具</a></li>
        
        <li><a href="../../../cn/cv-cn/shixiang">简历</a></li>
        
        <li><a href="../../../logo/qrcode.jpg">公众号</a></li>
        
        <li><a href="https://www.jianshu.com/u/b6608e27dc74">简书</a></li>
        
        <li><a href="../../../cn/todo">ToDo</a></li>
        
        <li><a href="../../../en/">English</a></li>
        
        

<li class="menu-extra"></li>



<li><a href="https://github.com/ShixiangWang/home/edit/master/content/cn%5cpost%5c2017-10-09-microArray-data-analysis.md" target="_blank">编辑</a></li>


<li><a href="../../../cn/index.xml" type="application/rss+xml" title="RSS feed">订阅</a></li>

<li><a href="http://creativecommons.org/licenses/by-nc-sa/4.0/" title="Attribution-NonCommercial-ShareAlike 4.0 International">版权</a></li>


        </ul>
      </nav>
    </header>

    <article class="main">
      <header class="title">
        

<h1>Bioconductor分析基因芯片数据</h1>



<h3>王诗翔 &middot 
2017-10-09</h3> 


   
  


      </header>


<p><em>参考学习《R语言与Bioconductor——生物信息学应用》第五章</em>，根据自己学习有删改，欢迎分享。本文中有些预备知识或一些方法的原理没有提及，请自行搜索理解或者购买该书学习。</p>
<p><strong>Bioconductor最初就是设计来分析基因芯片数据的，因此芯片分析整体反映了Biocondutor的设计理念和编程思想。</strong></p>
<h2 id="heading">快速入门</h2>
<p>通过一个例子了解如何载入芯片数据，完成预处理，最后获得基因（探针组）表达矩阵。</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-R" data-lang="R"><span style="color:#75715e"># install CLL package</span>
<span style="color:#75715e"># library(BiocInstaller)</span>
<span style="color:#75715e"># biocLite(&#34;CLL&#34;)</span>

<span style="color:#75715e"># 载入CLL包，CLL包会自动调用affy包，该包含有一系列处理函数</span>
<span style="color:#a6e22e">library</span>(CLL)
<span style="color:#75715e"># read example dataset，（CLL包附带的示例数据集）</span>
<span style="color:#a6e22e">data</span>(<span style="color:#e6db74">&#34;</span><span style="color:#e6db74">CLLbatch&#34;</span>)
<span style="color:#75715e"># pre-process using RMA method</span>
CLLrma <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">rma</span>(CLLbatch)
<span style="color:#75715e"># read expression value after pre-processing</span>
e <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">exprs</span>(CLLrma)
<span style="color:#75715e"># 查看部分数据</span>
e[1<span style="color:#f92672">:</span><span style="color:#ae81ff">5</span>, <span style="color:#ae81ff">1</span><span style="color:#f92672">:</span><span style="color:#ae81ff">5</span>]
</code></pre></div><!-- raw HTML omitted -->
<p>数据输出如下：</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-R" data-lang="R"><span style="color:#f92672">&gt;</span> e[1<span style="color:#f92672">:</span><span style="color:#ae81ff">5</span>, <span style="color:#ae81ff">1</span><span style="color:#f92672">:</span><span style="color:#ae81ff">5</span>]
          CLL10.CEL CLL11.CEL CLL12.CEL CLL13.CEL CLL14.CEL
<span style="color:#ae81ff">100</span>_g_at   <span style="color:#ae81ff">7.495697</span>  <span style="color:#ae81ff">7.945159</span>  <span style="color:#ae81ff">7.861043</span>  <span style="color:#ae81ff">7.990252</span>  <span style="color:#ae81ff">7.889718</span>
<span style="color:#ae81ff">1000</span>_at    <span style="color:#ae81ff">7.251120</span>  <span style="color:#ae81ff">8.298723</span>  <span style="color:#ae81ff">8.473531</span>  <span style="color:#ae81ff">8.131041</span>  <span style="color:#ae81ff">8.051400</span>
<span style="color:#ae81ff">1001</span>_at    <span style="color:#ae81ff">4.457362</span>  <span style="color:#ae81ff">4.517529</span>  <span style="color:#ae81ff">4.357791</span>  <span style="color:#ae81ff">4.651236</span>  <span style="color:#ae81ff">4.473598</span>
<span style="color:#ae81ff">1002</span>_f_at  <span style="color:#ae81ff">3.984693</span>  <span style="color:#ae81ff">3.981675</span>  <span style="color:#ae81ff">4.065171</span>  <span style="color:#ae81ff">4.132480</span>  <span style="color:#ae81ff">4.065022</span>
<span style="color:#ae81ff">1003</span>_s_at  <span style="color:#ae81ff">6.437046</span>  <span style="color:#ae81ff">6.200826</span>  <span style="color:#ae81ff">6.412218</span>  <span style="color:#ae81ff">6.313572</span>  <span style="color:#ae81ff">6.105120</span>
</code></pre></div><p>对象<code>e</code>是一个12625行、24列的基因表达矩阵。可以看到，我们仅用了不到10行代码就完成了整个芯片的预处理过程。</p>
<p><strong>关于探针组</strong></p>
<blockquote>
<p>一张基因芯片包含上百万的探针，它们被整齐有序地印刷再芯片上。探针组（probe set）来自于一个基因，通常由20对或者11对探针组成。每一对探针都由匹配探针（PM）和错配探针组成，称为探针对。</p>
<p>探针序列地来源叫做参考序列，通常来自于公开的核酸数据库。不同的芯片类型，探针组在参考序列中的分布不同。</p>
<p>这里有一个概念误区需要注意：芯片数据中的基因表达矩阵往往是以探针组而不是以基因为单位的，每行都对应一个探针组的表达量。我们一般通过ID映射才对应到探针组代表的基因，它们的关系常常为多个探针组对应一个基因。实际应用中则不太注意区分。</p>
</blockquote>
<p><strong>关于芯片文件格式</strong></p>
<blockquote>
<p>芯片实验结果获取数据主要通过两个步骤，第一步由扫描设备对芯片进行扫描，得到荧光信号图像文件（DAT文件）；第二步由系统自带的图形处理软件进行一系列图形处理与识别，从芯片图像中提取数据，得到CEL文件。</p>
<p>Affymetrix芯片原始数据最常用格式为CEL格式，也是芯片数据预处理和分析的出发点。</p>
<p>**CEL文件只提供了每个探针的灰度信息，还需要基因芯片探针排布的信息（哪个探针来自哪个探针组），才可以得到芯片上每个探针组对应的表达数据，这就需要CDF文件。**另一个重要的文件是Probe文件，它提供了探针的序列信息。</p>
<p>除了上述提到的几种文件格式，常见的还有EXP文件、CHP文件、TXT文件和RPT文件。TXT文件特指包含基因表达矩阵的TXT文件，CHP文件与TXT文件内容基本相同。</p>
</blockquote>
<h2 id="heading-1">数据预处理</h2>
<p>预处理的目的是<strong>将探针水平的数据转换成基因表达数据</strong>。主要的数据结构有AffyBatch类和ExpressionSet类，前者用于存储探针水平数据（相当于CEL文件内容），而后者用于存储表达水平的数据（基因表达矩阵）。</p>
<p>预处理通过质量控制，剔除不合格的芯片数据，只保留合格的进入下一步处理。然后通过标准化，将所有芯片数据中的基因表达值变换到一个可以比较的水平，用于后续分析。</p>
<h3 id="heading-2">数据输入</h3>
<p>实际应用中，常见情况是从CEL文件中获得数据。在上一个例子中，数据包CLL中得到的芯片数据基本等同于CEL文件。</p>
<p><strong>无论是数据包还是文件输入，读入的数据会存入一个AffyBatch类型的对象中。</strong></p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-R" data-lang="R"><span style="color:#75715e"># load package</span>
<span style="color:#75715e"># library(CLL)</span>

<span style="color:#75715e"># load data</span>
<span style="color:#a6e22e">data</span>(CLLbatch)

<span style="color:#75715e"># view the data type</span>
<span style="color:#a6e22e">data.class</span>(CLLbatch)

<span style="color:#75715e"># load status info of all samples</span>
<span style="color:#a6e22e">data</span>(disease)
<span style="color:#75715e"># view</span>
disease

<span style="color:#75715e"># see help of AffyBatch</span>
<span style="color:#a6e22e">help</span>(<span style="color:#e6db74">&#34;</span><span style="color:#e6db74">AffyBatch&#34;</span>)
</code></pre></div><p>看一下输出的一些结果：</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-R" data-lang="R"><span style="color:#f92672">&gt;</span> <span style="color:#a6e22e">data.class</span>(CLLbatch)
[1] <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">AffyBatch&#34;</span>
<span style="color:#f92672">&gt;</span> disease
   SampleID  Disease
<span style="color:#ae81ff">1</span>     CLL10     <span style="color:#f92672">&lt;</span><span style="color:#66d9ef">NA</span><span style="color:#f92672">&gt;</span>
<span style="color:#ae81ff">2</span>     CLL11 progres.
<span style="color:#ae81ff">3</span>     CLL12   stable
<span style="color:#ae81ff">4</span>     CLL13 progres.
<span style="color:#ae81ff">5</span>     CLL14 progres.
<span style="color:#ae81ff">6</span>     CLL15 progres.
<span style="color:#ae81ff">7</span>     CLL16 progres.
<span style="color:#ae81ff">8</span>     CLL17   stable
<span style="color:#ae81ff">9</span>     CLL18   stable
<span style="color:#ae81ff">10</span>    CLL19 progres.
<span style="color:#ae81ff">11</span>     CLL1   stable
<span style="color:#ae81ff">12</span>    CLL20   stable
<span style="color:#ae81ff">13</span>    CLL21 progres.
<span style="color:#ae81ff">14</span>    CLL22   stable
<span style="color:#ae81ff">15</span>    CLL23 progres.
<span style="color:#ae81ff">16</span>    CLL24   stable
<span style="color:#ae81ff">17</span>     CLL2   stable
<span style="color:#ae81ff">18</span>     CLL3 progres.
<span style="color:#ae81ff">19</span>     CLL4 progres.
<span style="color:#ae81ff">20</span>     CLL5 progres.
<span style="color:#ae81ff">21</span>     CLL6 progres.
<span style="color:#ae81ff">22</span>     CLL7 progres.
<span style="color:#ae81ff">23</span>     CLL8 progres.
<span style="color:#ae81ff">24</span>     CLL9   stable
</code></pre></div><blockquote>
<p>CLL数据集存储了24位癌症病人样本的数据，所有病人根据健康状态分为两组：稳定期（Stable）组和进展组（Progressive）。</p>
</blockquote>
<p><strong>AffyBatch类</strong></p>
<blockquote>
<p>Biobase软件包中，AffyBatch类是从一个更基础的类eSet类衍生来的。eSet类非常重要，它被写成了一个虚类，衍生出许多非常重要的类，包括ExpressionSet类，SnpSet类以及AffyBatch类等。eSet是Bioconductor为基因表达数据格式所定制的标准，因此非常有必要熟悉eSet及其衍生类。</p>
</blockquote>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-R" data-lang="R"><span style="color:#f92672">&gt;</span> <span style="color:#a6e22e">phenoData</span>(CLLbatch)
An object of class <span style="color:#e6db74">&#39;</span><span style="color:#e6db74">AnnotatedDataFrame&#39;</span>
  sampleNames<span style="color:#f92672">:</span> CLL10.CEL CLL11.CEL <span style="color:#66d9ef">...</span> <span style="color:#a6e22e">CLL9.CEL </span>(<span style="color:#ae81ff">24</span> total)
  varLabels<span style="color:#f92672">:</span> sample
  varMetadata<span style="color:#f92672">:</span> labelDescription
<span style="color:#f92672">&gt;</span> <span style="color:#a6e22e">featureData</span>(CLLbatch)
An object of class <span style="color:#e6db74">&#39;</span><span style="color:#e6db74">AnnotatedDataFrame&#39;</span><span style="color:#f92672">:</span> none
<span style="color:#f92672">&gt;</span> <span style="color:#a6e22e">protocolData</span>(CLLbatch)
An object of class <span style="color:#e6db74">&#39;</span><span style="color:#e6db74">AnnotatedDataFrame&#39;</span><span style="color:#f92672">:</span> none
<span style="color:#f92672">&gt;</span> <span style="color:#a6e22e">annotation</span>(CLLbatch)
[1] <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">hgu95av2&#34;</span>
<span style="color:#f92672">&gt;</span> exprs_matrix <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">assayData</span>(CLLbatch)[[1]]
<span style="color:#f92672">&gt;</span> exprs_matrix[1<span style="color:#f92672">:</span><span style="color:#ae81ff">5</span>, <span style="color:#ae81ff">1</span><span style="color:#f92672">:</span><span style="color:#ae81ff">5</span>]
  CLL10.CEL CLL11.CEL CLL12.CEL CLL13.CEL CLL14.CEL
<span style="color:#ae81ff">1</span>     <span style="color:#ae81ff">183.0</span>     <span style="color:#ae81ff">113.0</span>     <span style="color:#ae81ff">119.0</span>     <span style="color:#ae81ff">130.0</span>     <span style="color:#ae81ff">133.0</span>
<span style="color:#ae81ff">2</span>   <span style="color:#ae81ff">11524.8</span>    <span style="color:#ae81ff">6879.8</span>    <span style="color:#ae81ff">7891.3</span>    <span style="color:#ae81ff">8627.5</span>    <span style="color:#ae81ff">8205.3</span>
<span style="color:#ae81ff">3</span>     <span style="color:#ae81ff">301.0</span>     <span style="color:#ae81ff">146.0</span>     <span style="color:#ae81ff">133.0</span>     <span style="color:#ae81ff">160.0</span>     <span style="color:#ae81ff">153.0</span>
<span style="color:#ae81ff">4</span>   <span style="color:#ae81ff">11317.0</span>    <span style="color:#ae81ff">6747.0</span>    <span style="color:#ae81ff">7916.0</span>    <span style="color:#ae81ff">8616.0</span>    <span style="color:#ae81ff">7865.0</span>
<span style="color:#ae81ff">5</span>     <span style="color:#ae81ff">115.0</span>      <span style="color:#ae81ff">82.0</span>      <span style="color:#ae81ff">78.8</span>     <span style="color:#ae81ff">101.0</span>      <span style="color:#ae81ff">65.3</span>
<span style="color:#f92672">&gt;</span> <span style="color:#a6e22e">exprs</span>(CLLbatch)[1<span style="color:#f92672">:</span><span style="color:#ae81ff">5</span>, <span style="color:#ae81ff">1</span><span style="color:#f92672">:</span><span style="color:#ae81ff">5</span>]
  CLL10.CEL CLL11.CEL CLL12.CEL CLL13.CEL CLL14.CEL
<span style="color:#ae81ff">1</span>     <span style="color:#ae81ff">183.0</span>     <span style="color:#ae81ff">113.0</span>     <span style="color:#ae81ff">119.0</span>     <span style="color:#ae81ff">130.0</span>     <span style="color:#ae81ff">133.0</span>
<span style="color:#ae81ff">2</span>   <span style="color:#ae81ff">11524.8</span>    <span style="color:#ae81ff">6879.8</span>    <span style="color:#ae81ff">7891.3</span>    <span style="color:#ae81ff">8627.5</span>    <span style="color:#ae81ff">8205.3</span>
<span style="color:#ae81ff">3</span>     <span style="color:#ae81ff">301.0</span>     <span style="color:#ae81ff">146.0</span>     <span style="color:#ae81ff">133.0</span>     <span style="color:#ae81ff">160.0</span>     <span style="color:#ae81ff">153.0</span>
<span style="color:#ae81ff">4</span>   <span style="color:#ae81ff">11317.0</span>    <span style="color:#ae81ff">6747.0</span>    <span style="color:#ae81ff">7916.0</span>    <span style="color:#ae81ff">8616.0</span>    <span style="color:#ae81ff">7865.0</span>
<span style="color:#ae81ff">5</span>     <span style="color:#ae81ff">115.0</span>      <span style="color:#ae81ff">82.0</span>      <span style="color:#ae81ff">78.8</span>     <span style="color:#ae81ff">101.0</span>      <span style="color:#ae81ff">65.3</span>
</code></pre></div><ul>
<li>assayData槽是AffyBatch类必不可少的，其方法返回结果的第一个元素是矩阵类型，用于保存基因表达矩阵。该矩阵的行对应不同的探针组，用一个无重复的索引值表示；列对应不同的样本。当使用exprs方法时，调取的就是这个基因表达矩阵。</li>
<li>experimentData槽是一个MIAME（Minimum Information About a Microarray Experiment）类型的数据，设计这个MIAME类的目的就是用于保存MIAME原则建议的注释信息。该原则建议了一组标准用来记录与基因芯片实验设计相关的资料。</li>
</ul>
<p>结果中可以看到<code>phenoData(CLLbatch)</code>结果没有每个样品对应的表型信息。因为CLL包采用了一个数据框类型的变量disease来保存每个样本的表型信息。其实，将表型信息存入数据结构CLLbatch中是更为规范的做法。</p>
<h3 id="heading-3">质量控制</h3>
<p>这里的质量控制主要集中在CEL文件级别的处理，<strong>从最简单的直接观察，到平均值方法，再到比较高级的数据拟合方法</strong>。这三个层次的质量控制功能分别由image函数、simpleaffy包和affyPLM包实现。</p>
<p>我们先直观地看一下芯片上所有位点的灰度图像。</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-R" data-lang="R"><span style="color:#75715e"># 查看第一张芯片的灰度图像</span>
<span style="color:#a6e22e">image</span>(CLLbatch[,<span style="color:#ae81ff">1</span>])
</code></pre></div><p><img src="../../../figures/cel.png" alt=""></p>
<p>芯片左上角的花纹和芯片名称可以帮助我们来了解芯片数据是否可靠。根据图像信息，可以对芯片的信号强度产生一个总体认识：图像特别黑，信号强度低；图像特别亮，信号可能过饱和。</p>
<p>很显然这种方法非常粗糙，不能量化。比它好的简单方法是基于各种平均值的方法。<strong>这类方法的一个共同特点就是假设一组实验中的每个芯片数据对于某个平均值指标都相差不大</strong>。Affymetrix公司在指导手册中进行了详细描述。</p>
<ul>
<li>尺度因子（Scaling factor）：用于比较地芯片之间地尺度因子地比例必须小于3</li>
<li>检测值（Detection call）和检出率（Percent present）：一组探针能否被检测到，用检测值有（Present，简称P）、无（Absent，简称A）和不确定（Marginal present，简称M）来表示。</li>
<li>平均背景噪声（Average background）</li>
<li>标准内参（Internal control genes）:大部分的细胞有$\beta$-actin和GAPDH基因&hellip;</li>
</ul>
<p>根据上述的这些标准，使用Bioconductor的simpleaffy包对Affymetrix芯片数据进行质量评估，最后得到质量控制总览图。</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-R" data-lang="R"><span style="color:#75715e">## use simpleaffy package</span>
<span style="color:#a6e22e">library</span>(BiocInstaller)
<span style="color:#a6e22e">biocLite</span>(<span style="color:#e6db74">&#34;</span><span style="color:#e6db74">simpleaffy&#34;</span>)
<span style="color:#a6e22e">library</span>(simpleaffy)
<span style="color:#a6e22e">library</span>(CLL)
<span style="color:#a6e22e">data</span>(CLLbatch)
<span style="color:#75715e"># acquire analysis report</span>
Data.qc <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">qc</span>(CLLbatch)
<span style="color:#75715e"># visualize report</span>
<span style="color:#a6e22e">plot</span>(Data.qc)
</code></pre></div><p><img src="../../../figures/qc.png" alt=""></p>
<p>上图显示的是全部数据的质量控制总览图。第一列是所有样本的名称；第二列是两个数字，上面的是以百分比形式出现的检出率，下面的数字表明平均背景噪声。第三列（“QC Stats”）最下面的横轴是尺度因子等指标对应的坐标，取值范围从-3到3。第3列用到了三项指标：尺度因子、GAPDH 3&rsquo;/5&rsquo; 比值和actin 3&rsquo;/5'比值，分别用实心圆、空心圆和三角标志表示出来。如果出现红色的“bioB”字样，说明该样品中未能检测到BioB。</p>
<p>简单地讲，<strong>所有指标出现蓝色表示正常，红色表示可能存在质量问题</strong>。但这样根据实际情况进一步分析。一般来讲，如果有一个芯片各项指标都不太正常，尤其是BioB无法检测到，建议判定该芯片实验失败。比如CLL15.CEL这个数据的检出率明显低于其他样品，actin3/actin5远大于3，而且没有检测到BioB，因此数据可以判定为无效。</p>
<p><strong>基于平均值假设的评价指标都有一个默认的假设，那就是对于每一块芯片，质量是均匀的，不会随位置不同发生较大的变化</strong>。但事实上，芯片的每一个小格之间的质量也是有差异的。</p>
<p>如何得到比较可靠的质量评估呢？这需要设计多种能反映芯片数据全貌的指标综合分析从而得到最终的结论。<strong>这些指标要在对原始数据拟合的基础上计算得到</strong>，以图形显示，包括权重残差图、相对对数表达箱线图、相对标准差箱线图、RNA降解曲线、聚类分析、主成分分析、信号强度分布图与MA图等等。以上功能可以通过affyPLM包实现。</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-R" data-lang="R"><span style="color:#75715e">## use affyPLM package</span>
<span style="color:#a6e22e">biocLite</span>(<span style="color:#e6db74">&#34;</span><span style="color:#e6db74">affyPLM&#34;</span>)
<span style="color:#a6e22e">library</span>(affyPLM)
<span style="color:#a6e22e">library</span>(CLL)
<span style="color:#a6e22e">data</span>(CLLbatch)
<span style="color:#75715e"># 对数据集做回归分析，结果为一个PLMset类型的对象</span>
Pset <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">fitPLM</span>(CLLbatch)
<span style="color:#a6e22e">image</span>(CLLbatch[,<span style="color:#ae81ff">1</span>])
<span style="color:#75715e"># 根据计算结果，画权重图</span>
<span style="color:#a6e22e">image</span>(Pset, type<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;</span><span style="color:#e6db74">weights&#34;</span>, which<span style="color:#f92672">=</span><span style="color:#ae81ff">1</span>, main<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;</span><span style="color:#e6db74">Weights&#34;</span>)
<span style="color:#75715e"># 根据计算结果，画残差图</span>
<span style="color:#a6e22e">image</span>(Pset,  type<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;</span><span style="color:#e6db74">resids&#34;</span>, which<span style="color:#f92672">=</span><span style="color:#ae81ff">1</span>, main<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;</span><span style="color:#e6db74">Residuals&#34;</span>)
<span style="color:#75715e"># 根据计算结果，画残差符号图</span>
<span style="color:#a6e22e">image</span>(Pset, type<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;</span><span style="color:#e6db74">sign.resids&#34;</span>, which<span style="color:#f92672">=</span><span style="color:#ae81ff">1</span>, main<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;</span><span style="color:#e6db74">Residuals.sign&#34;</span>)
</code></pre></div><p>affyPLM包在探针水平拟合时引入了加权最小二乘法来进行回归，理论上，权重和残差 的分布是随机的，应该看到绿色均匀分布的权重图和红蓝均匀分布的残差图。</p>
<p>相对对数表达箱线图（RLE）反映了基因表达量的一致性趋势，它<strong>定义为一个探针组在某个样品的表达值除以该探针组在所有样品中表达值的中位数后取对数</strong>。</p>
<p>如果使用RLE箱线图来控制CLL数据集的实验质量，<strong>每个样品的中心应该非常接近纵坐标0的位置</strong>。如果个别样品的表现与其他大多数明显不同，那说明可能这个样品可能有问题。下面我们根据这个数据集来看看。</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-R" data-lang="R"><span style="color:#75715e">## </span>
<span style="color:#a6e22e">biocLite</span>(<span style="color:#e6db74">&#34;</span><span style="color:#e6db74">RColorBrewer&#34;</span>)
<span style="color:#75715e"># library(affyPLM)</span>
<span style="color:#a6e22e">library</span>(RColorBrewer)
colors <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">brewer.pal</span>(<span style="color:#ae81ff">12</span>, <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">Set3&#34;</span>)
<span style="color:#75715e"># RLE</span>
<span style="color:#a6e22e">Mbox</span>(Pset, ylim<span style="color:#f92672">=</span><span style="color:#a6e22e">c</span>(<span style="color:#ae81ff">-1</span>, <span style="color:#ae81ff">1</span>), col<span style="color:#f92672">=</span>colors, main<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;</span><span style="color:#e6db74">RLE&#34;</span>, las<span style="color:#f92672">=</span><span style="color:#ae81ff">3</span>)
<span style="color:#75715e"># NUSE</span>
<span style="color:#a6e22e">boxplot</span>(Pset, ylim<span style="color:#f92672">=</span><span style="color:#a6e22e">c</span>(<span style="color:#ae81ff">0.95</span>, <span style="color:#ae81ff">1.22</span>), col<span style="color:#f92672">=</span>colors, main<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;</span><span style="color:#e6db74">NUSE&#34;</span>, las<span style="color:#f92672">=</span><span style="color:#ae81ff">3</span>)
</code></pre></div><p><img src="../../../figures/rle.png" alt=""></p>
<p>NUSE是一种比RLE更为敏感的质量检测手段。我们可以结合NUSE图来确定是否某个芯片质量有问题。NUSE定义为一个探针组在某个样品的PM值的标准差除以该探针值在各样品中的PM标准差的中位数。<strong>如果所有芯片的质量都是非常可靠的话，那么它们的标准差会非常接近，因此它们的NUSE值都会在1附近</strong>。</p>
<blockquote>
<p>然而，如果有些芯片质量有问题的话，就会严重地偏离1，进而导致其他芯片地NUSE值偏向相反的方向。当然，存在一种极端的情况，就是大部分芯片都出问题了，但是它们的标准差却比较接近，反而会显得没有质量问题的芯片的NUSE值会明显偏离1，所以必须结合两种图作出可靠的判断。</p>
</blockquote>
<p><img src="../../../figures/nuse.png" alt=""></p>
<p>可以看到CLL1和CLL10的质量明显有别于其他样品，需要舍弃。</p>
<p><strong>RNA降解是影响芯片质量的一个很重要因素</strong>。因为RNA是从5’端开始降解的，理论上5‘端的荧光强度应该低于3’端的荧光强度。如果RNA降解曲线的斜率太小，甚至接近于0，很可能是RNA降解太严重，需要作为坏数据去除（下图显示的CLL13）。</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-R" data-lang="R"><span style="color:#75715e"># RNA degradation</span>
<span style="color:#75715e"># library(affy)</span>
data.deg <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">AffyRNAdeg</span>(CLLbatch)
<span style="color:#a6e22e">plotAffyRNAdeg</span>(data.deg, col<span style="color:#f92672">=</span>colors)
<span style="color:#a6e22e">legend</span>(<span style="color:#e6db74">&#34;</span><span style="color:#e6db74">topleft&#34;</span>, <span style="color:#a6e22e">rownames</span>(<span style="color:#a6e22e">pData</span>(CLLbatch)), col<span style="color:#f92672">=</span>colors, lwd<span style="color:#f92672">=</span><span style="color:#ae81ff">1</span>, inset<span style="color:#f92672">=</span><span style="color:#ae81ff">0.05</span>, cex<span style="color:#f92672">=</span><span style="color:#ae81ff">0.5</span>)
</code></pre></div><p><img src="../../../figures/RNAdeg.png" alt=""></p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-R" data-lang="R">CLLbatch <span style="color:#f92672">&lt;-</span> CLLbatch[, <span style="color:#f92672">-</span><span style="color:#a6e22e">match</span>(<span style="color:#a6e22e">c</span>(<span style="color:#e6db74">&#34;</span><span style="color:#e6db74">CLL10.CEL&#34;</span>, <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">CLL1.CEL&#34;</span>, <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">CLL13.CEL&#34;</span>),
                              <span style="color:#a6e22e">sampleNames</span>(CLLbatch))]
</code></pre></div><p>去掉三个质量差的。</p>
<p>前面提到的质量控制方法都是<strong>基于平均值思想</strong>的，其实，我们还可以从芯片之间的相互关系来检验芯片的质量。理论上组内同种类型的芯片数据应该聚拢在一起，两个组之间应该明显地分离。Pearson线性相关系数就是最常用的这类指标，但在实际应用中，往往不是直接查看相关系数矩阵，而是根据由相关系数矩阵导出的<strong>距离矩阵</strong>，进行聚类分析或主成分分析以对样品归类并图形化显示。</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-R" data-lang="R"><span style="color:#75715e">## cluster and PCA</span>
<span style="color:#75715e"># install packages</span>
<span style="color:#a6e22e">biocLite</span>(<span style="color:#e6db74">&#34;</span><span style="color:#e6db74">gcrma&#34;</span>)
<span style="color:#a6e22e">biocLite</span>(<span style="color:#e6db74">&#34;</span><span style="color:#e6db74">graph&#34;</span>)
<span style="color:#a6e22e">biocLite</span>(<span style="color:#e6db74">&#34;</span><span style="color:#e6db74">affycoretools&#34;</span>)

<span style="color:#75715e"># load packages</span>
<span style="color:#a6e22e">library</span>(CLL)
<span style="color:#a6e22e">library</span>(gcrma)
<span style="color:#a6e22e">library</span>(graph)
<span style="color:#a6e22e">library</span>(affycoretools)

<span style="color:#75715e"># load data</span>
<span style="color:#a6e22e">data</span>(CLL)
<span style="color:#a6e22e">data</span>(disease)
<span style="color:#75715e"># pre-process using gcrma</span>
CLLgcrma <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">gcrma</span>(CLLbatch)
<span style="color:#75715e"># retrieve gene expression dataset</span>
eset <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">exprs</span>(CLLgcrma)
<span style="color:#75715e"># compute Pearson corr coeffience</span>
pearson_cor <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">cor</span>(eset)
dist.lower <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">as.dist</span>(<span style="color:#ae81ff">1</span> <span style="color:#f92672">-</span> pearson_cor)
<span style="color:#75715e"># cluster</span>
hc <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">hclust</span>(dist.lower, <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">ave&#34;</span>)
<span style="color:#75715e"># plot</span>
<span style="color:#a6e22e">plot</span>(hc)
<span style="color:#75715e"># PCA</span>
samplenames <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">sub</span>(pattern <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">\\.CEL&#34;</span>, replacement <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">&#34;</span>, <span style="color:#a6e22e">colnames</span>(eset))
groups <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">factor</span>(disease[,<span style="color:#ae81ff">2</span>])
<span style="color:#a6e22e">plotPCA</span>(eset, addtext<span style="color:#f92672">=</span>samplenames, groups<span style="color:#f92672">=</span>groups, groupnames<span style="color:#f92672">=</span><span style="color:#a6e22e">levels</span>(groups))
</code></pre></div><p><img src="../../../figures/cluster.png" alt=""></p>
<p>从聚类分析的结果来看，稳定组和恶化组根本不能很好分开。这还不能判定实验失败。**理论上讲，如果总体上两组数据是分开的，那么说明我们关心的导致癌症从稳定到恶化的因素起到了主要因素。如果不是，需要具体问题具体分析。**CLL数据的实验样本来自不同个体，而不是细胞，很可能个体差异起到了主导作用，因此导致聚类被整体打乱。所以只有当聚类图中有明显的类别差异时，才适合考虑去除个别不归类的样品；如果整体分类被打乱，则不能简单判定所有样品都出了问题。</p>
<p>芯片分析往往采用两个主成分来构建分类图，我们从下图可以看出两组根本就不能很好分开。</p>
<p><img src="../../../figures/pca.png" alt=""></p>
<p>注意，使用主成分分析时，还必须考虑前2个主成分是否具有代表性，这要看前2个主成分的累计贡献率，如果低于60%，可以考虑采用多维尺度分析来构建分类图。</p>
<h3 id="heading-4">背景校正、标准化和汇总</h3>
<p>芯片数据通过<strong>质量控制</strong>，剔除不合格的样品，留下的数据经过<strong>背景校正、标准化和汇总</strong>三步得到下一步分析用的基因表达矩阵。</p>
<p>实际中经常发现大量的MM（Mismatch）信号比PM（Perfect Match）信号还要高。我们需要应用复杂的统计模型来去除背景噪声，这叫作背景校正（Background correction）。</p>
<p><strong>标准化的目的</strong>是消除测量间的非实验误差，使得实验条件下的测量可以相互比较。芯片数据标准化根据基本假设分为“bulk normalization&quot;与&quot;control-based normalization&quot;两种。前者假定仅有一小部分基因表达值在不同条件下有差异，因此使用所有的基因表达值作为参考进行标准化；后者使用表达值被认为恒定不变的参考基因作为标准进行标准化。实际应用中一般只采用第一种。</p>
<p>最后，我们使用一定的统计方法<strong>将前面得到的荧光强度值从探针水平汇总到探侦组水平</strong>，这个过程被称为汇总（Summarization）。</p>
<p>上述三个步骤可以由affy包的expresso函数实现，我们可以设定不同的参数来指定具体采用的方法。</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-R" data-lang="R"><span style="color:#75715e"># 在前面分析的基础上</span>
eset.mas <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">expresso</span>(CLLbatch, bgcorrect.method <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">mas&#34;</span>, normalize.method <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">constant&#34;</span>, pmcorrect.method <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">mas&#34;</span>,
                     summary.method <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">mas&#34;</span>)
</code></pre></div><p>可以看到函数的调用非常简单，但参数很多，理解比较复杂。想要了解所有参数的含义，使用<code>help(expresso)</code>命令获取参数说明。</p>
<p>下面命令可以查看可选的参数：</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-R" data-lang="R"><span style="color:#f92672">&gt;</span> <span style="color:#a6e22e">bgcorrect.methods</span>()
[1] <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">bg.correct&#34;</span> <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">mas&#34;</span>        <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">none&#34;</span>       <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">rma&#34;</span>       
<span style="color:#f92672">&gt;</span> <span style="color:#a6e22e">normalize.methods</span>(CLLbatch)
 [1] <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">constant&#34;</span>           <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">contrasts&#34;</span>          <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">invariantset&#34;</span>       <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">loess&#34;</span>              <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">methods&#34;</span>            <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">qspline&#34;</span>           
 [7] <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">quantiles&#34;</span>          <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">quantiles.robust&#34;</span>   <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">quantiles.probeset&#34;</span> <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">scaling&#34;</span>           
<span style="color:#f92672">&gt;</span> <span style="color:#a6e22e">pmcorrect.methods</span>()
[1] <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">mas&#34;</span>        <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">methods&#34;</span>    <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">pmonly&#34;</span>     <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">subtractmm&#34;</span>
<span style="color:#f92672">&gt;</span> <span style="color:#a6e22e">express.summary.stat.methods</span>()
[1] <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">avgdiff&#34;</span>      <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">liwong&#34;</span>       <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">mas&#34;</span>          <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">medianpolish&#34;</span> <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">playerout&#34;</span>  
</code></pre></div><table>
<thead>
<tr>
<th>参数</th>
<th>说明</th>
</tr>
</thead>
<tbody>
<tr>
<td>afbatch</td>
<td>输入数据的类型必须是AffyBatch对象</td>
</tr>
<tr>
<td>bgcorrect.method</td>
<td>背景校正方法</td>
</tr>
<tr>
<td>bgcorrect.param</td>
<td>指定背景校正方法的参数</td>
</tr>
<tr>
<td>normalize.method</td>
<td>标准化方法</td>
</tr>
<tr>
<td>normalize.param</td>
<td>指定标准化方法的参数</td>
</tr>
<tr>
<td>pmcorrect.method</td>
<td>PM调整方法</td>
</tr>
<tr>
<td>pmcorrect.param</td>
<td>指定PM方法参数</td>
</tr>
<tr>
<td>summary.method</td>
<td>汇总方法</td>
</tr>
<tr>
<td>summary.param</td>
<td>指定汇总方法的参数</td>
</tr>
</tbody>
</table>
<p><del>八种</del>十种标准方法，又可以分为芯片间标准方法和芯片内标准化方法。<strong>芯片间标准化方法针对单通道芯片数据</strong>，常用的有三种：<strong>线性缩放（constant）、不变集（invariantset）和分位数方法（quantiles）</strong>。<strong>芯片间标准方法</strong>的核心思想就是确定一个参考芯片（可以是假想的），假定芯片之间的某种不变量，对其他芯片数据进行整体的拉伸或者压缩变换。<strong>线性缩放</strong>方法，以其他芯片和参考芯片（默认为第一个）所有基因表达值均值的比值为因子，对其他芯片的表达值做等比例缩放。这个方法非常简单，是MAS5预处理算法默认使用的标准化方法。MAS5预处理算法不指定任何芯片，而是设定了一个假想的芯片均值作为参考，这样每个芯片都可以单独计算，有新的数据加入时也可以不必重新计算已经标准化的数据。<strong>分位数</strong>方法是RMA预处理算法默认使用的标准化方法（关于方法的使用应该根据数据合理选择，网上搜索方法使用的具体详情，这里书里没有进行过多的阐述）。</p>
<p>下面用MAS方法进行背景校正和标准化：</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-R" data-lang="R"><span style="color:#75715e"># 使用mas方法做背景校正</span>
<span style="color:#f92672">&gt;</span> CLLmas5 <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">bg.correct</span>(CLLbatch, method <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">mas&#34;</span>)
<span style="color:#75715e"># 使用constant方法标准化</span>
<span style="color:#f92672">&gt;</span> data_mas5 <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">normalize</span>(CLLmas5, method<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;</span><span style="color:#e6db74">constant&#34;</span>)
<span style="color:#75715e">#　查看每个样品的缩放系数</span>
<span style="color:#f92672">&gt;</span> <span style="color:#a6e22e">head</span>(<span style="color:#a6e22e">pm</span>(data_mas5)<span style="color:#f92672">/</span><span style="color:#a6e22e">pm</span>(CLLmas5), <span style="color:#ae81ff">5</span>)
       CLL11.CEL CLL12.CEL CLL14.CEL CLL15.CEL CLL16.CEL CLL17.CEL CLL18.CEL CLL19.CEL CLL20.CEL CLL21.CEL CLL22.CEL CLL23.CEL
<span style="color:#ae81ff">175218</span>         <span style="color:#ae81ff">1</span>  <span style="color:#ae81ff">1.155849</span>  <span style="color:#ae81ff">1.023873</span>  <span style="color:#ae81ff">1.493193</span>  <span style="color:#ae81ff">1.549369</span>  <span style="color:#ae81ff">2.000299</span>  <span style="color:#ae81ff">1.451576</span>  <span style="color:#ae81ff">1.776501</span> <span style="color:#ae81ff">0.9825108</span> <span style="color:#ae81ff">0.7070828</span> <span style="color:#ae81ff">0.9958733</span>  <span style="color:#ae81ff">1.432365</span>
<span style="color:#ae81ff">356689</span>         <span style="color:#ae81ff">1</span>  <span style="color:#ae81ff">1.155849</span>  <span style="color:#ae81ff">1.023873</span>  <span style="color:#ae81ff">1.493193</span>  <span style="color:#ae81ff">1.549369</span>  <span style="color:#ae81ff">2.000299</span>  <span style="color:#ae81ff">1.451576</span>  <span style="color:#ae81ff">1.776501</span> <span style="color:#ae81ff">0.9825108</span> <span style="color:#ae81ff">0.7070828</span> <span style="color:#ae81ff">0.9958733</span>  <span style="color:#ae81ff">1.432365</span>
<span style="color:#ae81ff">227696</span>         <span style="color:#ae81ff">1</span>  <span style="color:#ae81ff">1.155849</span>  <span style="color:#ae81ff">1.023873</span>  <span style="color:#ae81ff">1.493193</span>  <span style="color:#ae81ff">1.549369</span>  <span style="color:#ae81ff">2.000299</span>  <span style="color:#ae81ff">1.451576</span>  <span style="color:#ae81ff">1.776501</span> <span style="color:#ae81ff">0.9825108</span> <span style="color:#ae81ff">0.7070828</span> <span style="color:#ae81ff">0.9958733</span>  <span style="color:#ae81ff">1.432365</span>
<span style="color:#ae81ff">237919</span>         <span style="color:#ae81ff">1</span>  <span style="color:#ae81ff">1.155849</span>  <span style="color:#ae81ff">1.023873</span>  <span style="color:#ae81ff">1.493193</span>  <span style="color:#ae81ff">1.549369</span>  <span style="color:#ae81ff">2.000299</span>  <span style="color:#ae81ff">1.451576</span>  <span style="color:#ae81ff">1.776501</span> <span style="color:#ae81ff">0.9825108</span> <span style="color:#ae81ff">0.7070828</span> <span style="color:#ae81ff">0.9958733</span>  <span style="color:#ae81ff">1.432365</span>
<span style="color:#ae81ff">275173</span>         <span style="color:#ae81ff">1</span>  <span style="color:#ae81ff">1.155849</span>  <span style="color:#ae81ff">1.023873</span>  <span style="color:#ae81ff">1.493193</span>  <span style="color:#ae81ff">1.549369</span>  <span style="color:#ae81ff">2.000299</span>  <span style="color:#ae81ff">1.451576</span>  <span style="color:#ae81ff">1.776501</span> <span style="color:#ae81ff">0.9825108</span> <span style="color:#ae81ff">0.7070828</span> <span style="color:#ae81ff">0.9958733</span>  <span style="color:#ae81ff">1.432365</span>
       CLL24.CEL CLL2.CEL  CLL3.CEL  CLL4.CEL  CLL5.CEL CLL6.CEL CLL7.CEL CLL8.CEL  CLL9.CEL
<span style="color:#ae81ff">175218</span>  <span style="color:#ae81ff">1.706026</span> <span style="color:#ae81ff">1.156378</span> <span style="color:#ae81ff">0.8425419</span> <span style="color:#ae81ff">0.9775082</span> <span style="color:#ae81ff">0.9816573</span> <span style="color:#ae81ff">1.182963</span> <span style="color:#ae81ff">1.114976</span>  <span style="color:#ae81ff">1.13639</span> <span style="color:#ae81ff">0.8939248</span>
<span style="color:#ae81ff">356689</span>  <span style="color:#ae81ff">1.706026</span> <span style="color:#ae81ff">1.156378</span> <span style="color:#ae81ff">0.8425419</span> <span style="color:#ae81ff">0.9775082</span> <span style="color:#ae81ff">0.9816573</span> <span style="color:#ae81ff">1.182963</span> <span style="color:#ae81ff">1.114976</span>  <span style="color:#ae81ff">1.13639</span> <span style="color:#ae81ff">0.8939248</span>
<span style="color:#ae81ff">227696</span>  <span style="color:#ae81ff">1.706026</span> <span style="color:#ae81ff">1.156378</span> <span style="color:#ae81ff">0.8425419</span> <span style="color:#ae81ff">0.9775082</span> <span style="color:#ae81ff">0.9816573</span> <span style="color:#ae81ff">1.182963</span> <span style="color:#ae81ff">1.114976</span>  <span style="color:#ae81ff">1.13639</span> <span style="color:#ae81ff">0.8939248</span>
<span style="color:#ae81ff">237919</span>  <span style="color:#ae81ff">1.706026</span> <span style="color:#ae81ff">1.156378</span> <span style="color:#ae81ff">0.8425419</span> <span style="color:#ae81ff">0.9775082</span> <span style="color:#ae81ff">0.9816573</span> <span style="color:#ae81ff">1.182963</span> <span style="color:#ae81ff">1.114976</span>  <span style="color:#ae81ff">1.13639</span> <span style="color:#ae81ff">0.8939248</span>
<span style="color:#ae81ff">275173</span>  <span style="color:#ae81ff">1.706026</span> <span style="color:#ae81ff">1.156378</span> <span style="color:#ae81ff">0.8425419</span> <span style="color:#ae81ff">0.9775082</span> <span style="color:#ae81ff">0.9816573</span> <span style="color:#ae81ff">1.182963</span> <span style="color:#ae81ff">1.114976</span>  <span style="color:#ae81ff">1.13639</span> <span style="color:#ae81ff">0.8939248</span>
<span style="color:#75715e"># 查看第二个样品的缩放系数是怎么计算来的</span>
<span style="color:#f92672">&gt;</span> <span style="color:#a6e22e">mean</span>(<span style="color:#a6e22e">intensity</span>(CLLmas5)[,<span style="color:#ae81ff">1</span>])<span style="color:#f92672">/</span><span style="color:#a6e22e">mean</span>(<span style="color:#a6e22e">intensity</span>(CLLmas5)[,<span style="color:#ae81ff">2</span>])
[1] <span style="color:#ae81ff">1.155849</span>
</code></pre></div><p>芯片内标准化方法针对<strong>双通道芯片</strong>数据，又可以分为<strong>全局化方法</strong>（Global normalization）和<strong>荧光强度依赖的方法</strong>（Intensity-dependent normalization）。</p>
<p>全局化方法假设红色染料的信号强度与绿色染料的信号强度是正比例关系的，即R=kG（R：红色信号强度；G：绿色信号强度）。</p>
<p><img src="https://upload.wikimedia.org/wikipedia/commons/thumb/6/63/Pre-normalization.MAplot.png/450px-Pre-normalization.MAplot.png" alt=""></p>
<p><img src="https://upload.wikimedia.org/wikipedia/commons/thumb/3/3e/Post-normalization.MAplot.png/450px-Post-normalization.MAplot.png" alt=""></p>
<p><a href="https://en.wikipedia.org/wiki/MA_plot">MA（M代表Minus，A代表Average）图</a>全称为：The distribution of the red/green intensity ratio plotted by the average intensity。定义$M=log_2(R/G)，A=1/2<em>log_2(R</em>G)$。此处R和G已经不再特指双通道cDNA芯片中红和绿标记的样品表达量，可以表示任何两个需要对比的数据。<strong>单通道数据中MA图反应了基因在对比的样品中表达差异（对数化的）随基因信号强度（对数化的）变化的分布</strong>。</p>
<p>当数据符合全局化方法的假设，<strong>数据标准化后的MA图大多数基因的差异表达值（M值）应该对称分布在水平的中心线（M=0）附近</strong>。若此方法不成立，只能<strong>采用荧光强度依赖的方法，将M调整为以0为中心的分布</strong>，这类方法最常用的是Loess方法。简单来说就是对不同的A值的基因进行局部加权回归，得到一条蓝色的直线。不过现在已经很少有人用双通道cDNA芯片了。</p>
<h3 id="heading-5">预处理的一体化算法</h3>
<p>除了expresso函数，affyPLM软件包提供了threestep函数可以更快地实现同样的功能。如果我们自己自行组合参数，实际上使用可能会很低效，甚至有些组合不能使用。因此实际中，应用较多的预设参数的一体化算法，我们可以完全不管细节，直接使用从而简化处理和分析过程。</p>
<p><strong>预处理方法</strong></p>
<table>
<thead>
<tr>
<th align="center">方法</th>
<th align="center">背景校正方法</th>
<th align="center">标准化方法</th>
<th align="center">汇总方法</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center">MASS</td>
<td align="center">mas</td>
<td align="center">constant</td>
<td align="center">mas</td>
</tr>
<tr>
<td align="center">dChip</td>
<td align="center">mas</td>
<td align="center">invariantset</td>
<td align="center">liwong</td>
</tr>
<tr>
<td align="center">RMA</td>
<td align="center">rma</td>
<td align="center">quantile</td>
<td align="center">medianpolish</td>
</tr>
</tbody>
</table>
<p>常见的预处理算法一般都已经按照同名函数包装好以供调用，如affy包的MAS5和RMA，gcrma包的gcRMA（基于RMA算法）等。</p>
<p>实际工作中，MAS5和RMA算法使用较多。下面对差异做一个比较：</p>
<ul>
<li>MAS5 每个芯片可以单独进行标准化；RMA 采用多芯片模型，需要对所有芯片一起进行标准化。</li>
<li>MAS5 利用MM探针的信息去除背景噪声，基本思路是MP-MM；RMA 不使用MM信息，而是基于PM信号分布采用的随机模型来估计表达值。</li>
<li><strong>RMA处理后的数据是经过2为底的对数转换的，而MAS5不是，这一点非常重要，因为大多数芯片分析软件和函数的输入数据必须经过对数变换。</strong></li>
</ul>
<p>下面我们来比较不同算法的处理效果。</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-R" data-lang="R"><span style="color:#a6e22e">library</span>(affy)
<span style="color:#a6e22e">library</span>(gcrma)
<span style="color:#a6e22e">library</span>(affyPLM)
<span style="color:#a6e22e">library</span>(RColorBrewer)
<span style="color:#a6e22e">library</span>(CLL)

<span style="color:#a6e22e">data</span>(<span style="color:#e6db74">&#34;</span><span style="color:#e6db74">CLLbatch&#34;</span>)
colors <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">brewer.pal</span>(<span style="color:#ae81ff">12</span>, <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">Set3&#34;</span>)
<span style="color:#75715e"># use MAS5</span>
CLLmas5 <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">mas5</span>(CLLbatch)
<span style="color:#75715e"># use rma </span>
CLLrma <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">rma</span>(CLLbatch)
<span style="color:#75715e"># use gcrma</span>
CLLgcrma <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">gcrma</span>(CLLbatch)

<span style="color:#75715e">## hist plot</span>
<span style="color:#a6e22e">hist</span>(CLLbatch, main<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;</span><span style="color:#e6db74">orignal&#34;</span>, col<span style="color:#f92672">=</span>colors)
<span style="color:#a6e22e">legend</span>(<span style="color:#e6db74">&#34;</span><span style="color:#e6db74">topright&#34;</span>, <span style="color:#a6e22e">rownames</span>(<span style="color:#a6e22e">pData</span>(CLLbatch)), col<span style="color:#f92672">=</span>colors,
       lwd<span style="color:#f92672">=</span><span style="color:#ae81ff">1</span>, inset<span style="color:#f92672">=</span><span style="color:#ae81ff">0.05</span>, cex<span style="color:#f92672">=</span><span style="color:#ae81ff">0.5</span>, ncol<span style="color:#f92672">=</span><span style="color:#ae81ff">3</span>)
<span style="color:#a6e22e">hist</span>(CLLmas5, main<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;</span><span style="color:#e6db74">MAS 5.0&#34;</span>, xlim<span style="color:#f92672">=</span><span style="color:#a6e22e">c</span>(<span style="color:#ae81ff">-150</span>,<span style="color:#ae81ff">2</span>^10), col<span style="color:#f92672">=</span>colors)
<span style="color:#a6e22e">hist</span>(CLLrma, main<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;</span><span style="color:#e6db74">RMA&#34;</span>, col<span style="color:#f92672">=</span>colors)
<span style="color:#a6e22e">hist</span>(CLLgcrma, main<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;</span><span style="color:#e6db74">gcRMA&#34;</span>, col<span style="color:#f92672">=</span>colors)
<span style="color:#75715e">## boxplot</span>
<span style="color:#a6e22e">boxplot</span>(CLLbatch, col<span style="color:#f92672">=</span>colors, las<span style="color:#f92672">=</span><span style="color:#ae81ff">3</span>, main<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;</span><span style="color:#e6db74">original&#34;</span>)
<span style="color:#a6e22e">boxplot</span>(CLLmas5, col<span style="color:#f92672">=</span>colors, las<span style="color:#f92672">=</span><span style="color:#ae81ff">3</span>, ylim<span style="color:#f92672">=</span><span style="color:#a6e22e">c</span>(<span style="color:#ae81ff">0</span>,<span style="color:#ae81ff">1000</span>), main<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;</span><span style="color:#e6db74">MAS 5.0&#34;</span>)
<span style="color:#a6e22e">boxplot</span>(CLLrma, col<span style="color:#f92672">=</span>colors, las<span style="color:#f92672">=</span><span style="color:#ae81ff">3</span>, main<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;</span><span style="color:#e6db74">RMA&#34;</span>)
<span style="color:#a6e22e">boxplot</span>(CLLgcrma, col<span style="color:#f92672">=</span>colors, las<span style="color:#f92672">=</span><span style="color:#ae81ff">3</span>, main<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;</span><span style="color:#e6db74">gcRMA&#34;</span>)
</code></pre></div><p>先看信号强度分布图</p>
<p><img src="../../../figures/intensity_original.png" alt=""></p>
<p><img src="../../../figures/intensity_mas5.png" alt=""></p>
<p><img src="../../../figures/intensity_rma.png" alt=""></p>
<p><img src="../../../figures/intensity_gcrma.png" alt=""></p>
<p>这里简要说明：RMA算法将多条曲线重合到了一起，有利于进一步的差异分析，但却出现了双峰现象，不符合高斯正态分布。很显然gcRMA算法在这里表现的更好。当然，这不意味着gcRMA算法总是优越于RMA算法。</p>
<p>下面看箱线图结果：</p>
<p><img src="../../../figures/boxplot_original.png" alt=""></p>
<p><img src="../../../figures/boxplot_mas.png" alt=""></p>
<p><img src="../../../figures/boxplot_rma.png" alt=""></p>
<p><img src="../../../figures/boxplot_gcrma.png" alt=""></p>
<p>三个算法处理后的各样品的中值都十分接近。MAS5算法总体而言还不错，有一定拖尾现象；而gcRMA的拖尾现象比RMA要明显得多。<strong>这说明针对低表达量的基因，RMA算法比gcRMA算法表现更好一些。</strong></p>
<p>最后我们可以通过MA图来查看标准化处理的效果。</p>
<p><img src="../../../figures/orignal_ma_plot.png" alt=""></p>
<p><img src="../../../figures/gcrma_ma_plot.png" alt=""></p>
<p>原始数据中，中值（红线）偏离0，经过gcRMA处理后，中值基本保持在零线上。</p>
<h2 id="heading-6">基因芯片数据分析</h2>
<p>通过以下几个实例，掌握Bioconductor处理芯片数据的完整过程。</p>
<h3 id="heading-7">选取差异表达基因</h3>
<p>这是非常关键也是为人所熟知的一步，即使没有数据分析背景的生物学方向学生也必然不会对此陌生。而生物实验往往想要找的就是差异表达基因。</p>
<p>总的来说，这类分析的<strong>基本假设是标准化的芯片数据符合正态分布</strong>（想想为什么正式分析数据前要做那么多步骤处理？），所用的统计方法基本上就是T/F检验和方差分析以及它们的改进方法。</p>
<p>这里不再详细介绍各类方法，现在经验贝叶斯是当前最常用的分析方法，它已经完整地由Bioconductor中的limma包实现。limma包是基于R和Bioconductor平台分析芯片数据的综合软件包，该包功能齐全、教程完善、使用率极高，几乎成了芯片数据处理流程的代名词（说明分析用它准没错了）。</p>
<p>下面是应用该包计算CLL数据集中差异表达基因的流程。</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-R" data-lang="R"><span style="color:#75715e">####</span>
<span style="color:#75715e">#### DEG analysis</span>

<span style="color:#75715e">#　如果没有安装limma包，请取消下面两行注释</span>
<span style="color:#75715e"># library(BiocInstaller)</span>
<span style="color:#75715e"># biocLite(&#34;limma&#34;)</span>

<span style="color:#75715e"># 导入包</span>
<span style="color:#a6e22e">library</span>(limma)
<span style="color:#a6e22e">library</span>(gcrma)
<span style="color:#a6e22e">library</span>(CLL)

<span style="color:#75715e"># 导入CLL数据</span>
<span style="color:#a6e22e">data</span>(<span style="color:#e6db74">&#34;</span><span style="color:#e6db74">CLLbatch&#34;</span>)
<span style="color:#a6e22e">data</span>(disease)

<span style="color:#75715e"># 移除 CLL1, CLL10, CLL13</span>
CLLbatch <span style="color:#f92672">&lt;-</span> CLLbatch[, <span style="color:#f92672">-</span><span style="color:#a6e22e">match</span>(<span style="color:#a6e22e">c</span>(<span style="color:#e6db74">&#34;</span><span style="color:#e6db74">CLL10.CEL&#34;</span>, <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">CLL1.CEL&#34;</span>, <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">CLL13.CEL&#34;</span>),
                              <span style="color:#a6e22e">sampleNames</span>(CLLbatch))]
<span style="color:#75715e"># 用gcRMA算法进行预处理</span>
CLLgcrma <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">gcrma</span>(CLLbatch)

<span style="color:#75715e"># remove .CEL in sample names</span>
<span style="color:#a6e22e">sampleNames</span>(CLLgcrma) <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">gsub</span>(<span style="color:#e6db74">&#34;</span><span style="color:#e6db74">.CEL$&#34;</span>, <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">&#34;</span>, <span style="color:#a6e22e">sampleNames</span>(CLLgcrma))
<span style="color:#75715e"># remove record in data disease about CLL1, 10 and 13.CEL</span>
disease <span style="color:#f92672">&lt;-</span> disease<span style="color:#a6e22e">[match</span>(<span style="color:#a6e22e">sampleNames</span>(CLLgcrma), disease[, <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">SampleID&#34;</span>]), ]

<span style="color:#75715e"># 获得余下21个样品的基因表达矩阵</span>
eset <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">exprs</span>(CLLgcrma)

<span style="color:#75715e"># 提取实验条件信息</span>
disease <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">factor</span>(disease[, <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">Disease&#34;</span>])

<span style="color:#75715e"># 构建实验设计矩阵</span>
design <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">model.matrix</span>(<span style="color:#f92672">~</span><span style="color:#ae81ff">-1</span><span style="color:#f92672">+</span>disease)
<span style="color:#75715e"># 构建对比模型，比较两个实验条件下表达数据</span>
<span style="color:#75715e">#　这里的.是简写而不是运算符号</span>
contrast.matrix <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">makeContrasts</span>(contrasts <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">diseaseprogres. - diseasestable&#34;</span>,
                                 levels<span style="color:#f92672">=</span>design)

<span style="color:#75715e"># 线性模型拟合</span>
fit <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">lmFit</span>(eset, design)
<span style="color:#75715e"># 根据对比模型进行差值计算 </span>
fit1 <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">contrasts.fit</span>(fit, contrast.matrix)
<span style="color:#75715e"># 贝叶斯检验</span>
fit2 <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">eBayes</span>(fit1)

<span style="color:#75715e"># 生成所有基因的检验结果报告</span>
dif <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">topTable</span>(fit2, coef<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;</span><span style="color:#e6db74">diseaseprogres. - diseasestable&#34;</span>, n<span style="color:#f92672">=</span><span style="color:#a6e22e">nrow</span>(fit2), lfc<span style="color:#f92672">=</span><span style="color:#a6e22e">log2</span>(<span style="color:#ae81ff">1.5</span>))
<span style="color:#75715e"># 用P.Value进行筛选，得到全部差异表达基因</span>
dif <span style="color:#f92672">&lt;-</span> dif[dif[, <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">P.Value&#34;</span>]<span style="color:#f92672">&lt;</span><span style="color:#ae81ff">0.01</span>,]
<span style="color:#75715e"># 显示一部分报告结果</span>
<span style="color:#a6e22e">head</span>(dif)
</code></pre></div><div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-R" data-lang="R"><span style="color:#f92672">&gt;</span> <span style="color:#a6e22e">head</span>(dif)
              logFC  AveExpr         t      P.Value adj.P.Val         B
<span style="color:#ae81ff">39400</span>_at <span style="color:#ae81ff">-0.9997850</span> <span style="color:#ae81ff">5.634004</span> <span style="color:#ae81ff">-5.727329</span> <span style="color:#ae81ff">1.482860e-05</span> <span style="color:#ae81ff">0.1034544</span> <span style="color:#ae81ff">2.4458354</span>
<span style="color:#ae81ff">1303</span>_at  <span style="color:#ae81ff">-1.3430306</span> <span style="color:#ae81ff">4.540225</span> <span style="color:#ae81ff">-5.596813</span> <span style="color:#ae81ff">1.974284e-05</span> <span style="color:#ae81ff">0.1034544</span> <span style="color:#ae81ff">2.1546350</span>
<span style="color:#ae81ff">33791</span>_at  <span style="color:#ae81ff">1.9647962</span> <span style="color:#ae81ff">6.837903</span>  <span style="color:#ae81ff">5.400499</span> <span style="color:#ae81ff">3.047498e-05</span> <span style="color:#ae81ff">0.1034544</span> <span style="color:#ae81ff">1.7135583</span>
<span style="color:#ae81ff">36131</span>_at  <span style="color:#ae81ff">0.9574214</span> <span style="color:#ae81ff">9.945334</span>  <span style="color:#ae81ff">5.367741</span> <span style="color:#ae81ff">3.277762e-05</span> <span style="color:#ae81ff">0.1034544</span> <span style="color:#ae81ff">1.6396223</span>
<span style="color:#ae81ff">37636</span>_at  <span style="color:#ae81ff">2.0534093</span> <span style="color:#ae81ff">5.478683</span>  <span style="color:#ae81ff">5.120519</span> <span style="color:#ae81ff">5.699454e-05</span> <span style="color:#ae81ff">0.1439112</span> <span style="color:#ae81ff">1.0788313</span>
<span style="color:#ae81ff">36122</span>_at  <span style="color:#ae81ff">0.8008604</span> <span style="color:#ae81ff">7.146293</span>  <span style="color:#ae81ff">4.776721</span> <span style="color:#ae81ff">1.241402e-04</span> <span style="color:#ae81ff">0.2612118</span> <span style="color:#ae81ff">0.2922106</span>
</code></pre></div><p>下面逐次介绍这个分析过程的<strong>六个</strong>关键步骤：<strong>构建基因表达矩阵、构建实验设计矩阵、构建对比模型（对比矩阵）、线性模型拟合、贝叶斯检验和生成结果报表</strong>。</p>
<p>构建基因表达矩阵时，需要注意limma包对输入数据要求是必须经过对数转换的表达值。如果使用的预处理算法没有该处理过程，我们需要自行编写代码处理。</p>
<p>实验设计矩阵需要调用model.matrix函数构建，该函数需要用户指定一个公式，构建好的实验设计矩阵design要提供给下一步的拟合函数lmFit。</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-R" data-lang="R"><span style="color:#f92672">&gt;</span> design
   diseaseprogres. diseasestable
<span style="color:#ae81ff">1</span>                <span style="color:#ae81ff">1</span>             <span style="color:#ae81ff">0</span>
<span style="color:#ae81ff">2</span>                <span style="color:#ae81ff">0</span>             <span style="color:#ae81ff">1</span>
<span style="color:#ae81ff">3</span>                <span style="color:#ae81ff">1</span>             <span style="color:#ae81ff">0</span>
<span style="color:#ae81ff">4</span>                <span style="color:#ae81ff">1</span>             <span style="color:#ae81ff">0</span>
<span style="color:#ae81ff">5</span>                <span style="color:#ae81ff">1</span>             <span style="color:#ae81ff">0</span>
<span style="color:#ae81ff">6</span>                <span style="color:#ae81ff">0</span>             <span style="color:#ae81ff">1</span>
<span style="color:#ae81ff">7</span>                <span style="color:#ae81ff">0</span>             <span style="color:#ae81ff">1</span>
<span style="color:#ae81ff">8</span>                <span style="color:#ae81ff">1</span>             <span style="color:#ae81ff">0</span>
<span style="color:#ae81ff">9</span>                <span style="color:#ae81ff">0</span>             <span style="color:#ae81ff">1</span>
<span style="color:#ae81ff">10</span>               <span style="color:#ae81ff">1</span>             <span style="color:#ae81ff">0</span>
<span style="color:#ae81ff">11</span>               <span style="color:#ae81ff">0</span>             <span style="color:#ae81ff">1</span>
<span style="color:#ae81ff">12</span>               <span style="color:#ae81ff">1</span>             <span style="color:#ae81ff">0</span>
<span style="color:#ae81ff">13</span>               <span style="color:#ae81ff">0</span>             <span style="color:#ae81ff">1</span>
<span style="color:#ae81ff">14</span>               <span style="color:#ae81ff">0</span>             <span style="color:#ae81ff">1</span>
<span style="color:#ae81ff">15</span>               <span style="color:#ae81ff">1</span>             <span style="color:#ae81ff">0</span>
<span style="color:#ae81ff">16</span>               <span style="color:#ae81ff">1</span>             <span style="color:#ae81ff">0</span>
<span style="color:#ae81ff">17</span>               <span style="color:#ae81ff">1</span>             <span style="color:#ae81ff">0</span>
<span style="color:#ae81ff">18</span>               <span style="color:#ae81ff">1</span>             <span style="color:#ae81ff">0</span>
<span style="color:#ae81ff">19</span>               <span style="color:#ae81ff">1</span>             <span style="color:#ae81ff">0</span>
<span style="color:#ae81ff">20</span>               <span style="color:#ae81ff">1</span>             <span style="color:#ae81ff">0</span>
<span style="color:#ae81ff">21</span>               <span style="color:#ae81ff">0</span>             <span style="color:#ae81ff">1</span>
<span style="color:#a6e22e">attr</span>(,<span style="color:#e6db74">&#34;</span><span style="color:#e6db74">assign&#34;</span>)
[1] <span style="color:#ae81ff">1</span> <span style="color:#ae81ff">1</span>
<span style="color:#a6e22e">attr</span>(,<span style="color:#e6db74">&#34;</span><span style="color:#e6db74">contrasts&#34;</span>)
<span style="color:#a6e22e">attr</span>(,<span style="color:#e6db74">&#34;</span><span style="color:#e6db74">contrasts&#34;</span>)<span style="color:#f92672">$</span>disease
[1] <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">contr.treatment&#34;</span>
</code></pre></div><p>实验设计矩阵的每一行对应一个样品的编码，每一列对应样品的一个特征。这里只考虑了一个因素两个水平，如果是多因素和多水平的实验设计，会产生更多的特征，需要参考文档设计。</p>
<p>比较模型需要调用makeContrasts函数构建，该函数需要用户指定一个公式，这个公式表明用户要求对实验矩阵design中的哪一列特征和哪一列特征进行比较，以得到差异。</p>
<p>接下来是根据实验设计矩阵调用函数对基因表达矩阵做线性拟合<code>lmFit(eset, design)</code>，根据对比模型进行差值计算，最后是贝叶斯检验。</p>
<p>topTable函数的使用需要注意几点：</p>
<ol>
<li>topTable提供了多种方法可以做基因筛选，本例就通过对数化的变化倍数“lfc”去掉了一些在两组条件下变化不大的基因，但是这样做的理由不够充分，因为变化倍数不大的不一定就是没有显著变化。</li>
<li>topTable还提供了参数可以对基因进行排序。</li>
<li>显著性基因的选取具有一定主观性，阈值设定是0.01还是0.05没有严格标准。</li>
</ol>
<p>此处分析步骤，还可以参考<a href="http://blog.sciencenet.cn/blog-295006-403640.html">【R高级教程】专题二：差异表达基因的分析</a>理解学习和实践。</p>
<h3 id="heading-8">注释</h3>
<p>找到差异表达基因之后，接下来是使用注释包对其进行注释（因为我们现在找到的是探针组）。</p>
<p>对Affymetrix芯片进行注释采用下载对应平台注释包进行本地注释的方式。下面例子只用两种基因ID来对探针组进行注释。</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-R" data-lang="R"><span style="color:#75715e"># 加载注释工具包</span>
<span style="color:#a6e22e">library</span>(annotate)

<span style="color:#75715e"># 获得基因芯片注释包名称</span>
affydb <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">annPkgName</span>(CLLbatch<span style="color:#f92672">@</span>annotation, type<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;</span><span style="color:#e6db74">db&#34;</span>)
affydb
<span style="color:#75715e"># 如果没有安装，先安装</span>
<span style="color:#a6e22e">biocLite</span>(affydb)
<span style="color:#75715e">#　加载该包</span>
<span style="color:#a6e22e">library</span>(affydb, character.only <span style="color:#f92672">=</span> <span style="color:#66d9ef">TRUE</span>)
<span style="color:#75715e"># 根据每个探针组的ID获取对应基因Gene Symbol，并作为新的一列</span>
dif<span style="color:#f92672">$</span>symbols <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">getSYMBOL</span>(<span style="color:#a6e22e">rownames</span>(dif), affydb)
<span style="color:#75715e"># 根据探针ID获取对应基因Entrez ID</span>
dif<span style="color:#f92672">$</span>EntrezID <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">getEG</span>(<span style="color:#a6e22e">rownames</span>(dif), affydb)
<span style="color:#75715e"># 显示前几行</span>
<span style="color:#a6e22e">head</span>(dif)
</code></pre></div><div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-R" data-lang="R"><span style="color:#f92672">&gt;</span> <span style="color:#a6e22e">head</span>(dif)
              logFC  AveExpr         t      P.Value adj.P.Val         B symbols EntrezID
<span style="color:#ae81ff">39400</span>_at <span style="color:#ae81ff">-0.9997850</span> <span style="color:#ae81ff">5.634004</span> <span style="color:#ae81ff">-5.727329</span> <span style="color:#ae81ff">1.482860e-05</span> <span style="color:#ae81ff">0.1034544</span> <span style="color:#ae81ff">2.4458354</span> TBC1D2B    <span style="color:#ae81ff">23102</span>
<span style="color:#ae81ff">1303</span>_at  <span style="color:#ae81ff">-1.3430306</span> <span style="color:#ae81ff">4.540225</span> <span style="color:#ae81ff">-5.596813</span> <span style="color:#ae81ff">1.974284e-05</span> <span style="color:#ae81ff">0.1034544</span> <span style="color:#ae81ff">2.1546350</span>  SH3BP2     <span style="color:#ae81ff">6452</span>
<span style="color:#ae81ff">33791</span>_at  <span style="color:#ae81ff">1.9647962</span> <span style="color:#ae81ff">6.837903</span>  <span style="color:#ae81ff">5.400499</span> <span style="color:#ae81ff">3.047498e-05</span> <span style="color:#ae81ff">0.1034544</span> <span style="color:#ae81ff">1.7135583</span>   DLEU1    <span style="color:#ae81ff">10301</span>
<span style="color:#ae81ff">36131</span>_at  <span style="color:#ae81ff">0.9574214</span> <span style="color:#ae81ff">9.945334</span>  <span style="color:#ae81ff">5.367741</span> <span style="color:#ae81ff">3.277762e-05</span> <span style="color:#ae81ff">0.1034544</span> <span style="color:#ae81ff">1.6396223</span>   CLIC1     <span style="color:#ae81ff">1192</span>
<span style="color:#ae81ff">37636</span>_at  <span style="color:#ae81ff">2.0534093</span> <span style="color:#ae81ff">5.478683</span>  <span style="color:#ae81ff">5.120519</span> <span style="color:#ae81ff">5.699454e-05</span> <span style="color:#ae81ff">0.1439112</span> <span style="color:#ae81ff">1.0788313</span>   JADE3     <span style="color:#ae81ff">9767</span>
<span style="color:#ae81ff">36122</span>_at  <span style="color:#ae81ff">0.8008604</span> <span style="color:#ae81ff">7.146293</span>  <span style="color:#ae81ff">4.776721</span> <span style="color:#ae81ff">1.241402e-04</span> <span style="color:#ae81ff">0.2612118</span> <span style="color:#ae81ff">0.2922106</span>    <span style="color:#f92672">&lt;</span><span style="color:#66d9ef">NA</span><span style="color:#f92672">&gt;</span>     <span style="color:#f92672">&lt;</span><span style="color:#66d9ef">NA</span><span style="color:#f92672">&gt;</span>
</code></pre></div><p><strong>注释实质上是一个ID映射过程</strong>，这里是将芯片探侦组的ID映射到基因国际标准名称（Gene symbol）和Entrez ID两种ID上。</p>
<p>Gene symbol是由人类基因命名委员会（The HUGO Gene Nomenclature Committee, HGNC）为每个人类基因提供的唯一命名，一般是大写拉丁字母缩写形式，后面可以加数字，以便于人工阅读。Entrez ID是NCBI的GI（GeneInfo Identifier）。NCBI对于每一条提交的序列，根据其存入NCBI数据时的先后顺序赋给一个整数，这就是GI。这里映射GI的目的是为了将GI映射到基因本体论（GO），然后做GO的富集分析。</p>
<h3 id="heading-9">统计分析及可视化</h3>
<p>对于差异表达分析，最主要的两种统计分析就是GO的富集分析和KEGG通路的富集分析。它们分别由GOstats包和GeneAnswers包实现。</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-R" data-lang="R"><span style="color:#75715e"># 加载包</span>
<span style="color:#a6e22e">library</span>(GOstats)
<span style="color:#75715e"># 提取HG_U95Av2芯片中所有探针组对应的EntrezID，注意保证uniq</span>
entrezUniverse <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">unique</span>(<span style="color:#a6e22e">unlist</span>(<span style="color:#a6e22e">mget</span>(<span style="color:#a6e22e">rownames</span>(eset), hgu95av2ENTREZID)))
<span style="color:#75715e"># 提取差异表达基因及其对应的EntrezID</span>
entrezSelected <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">unique</span>(dif[<span style="color:#f92672">!</span><span style="color:#a6e22e">is.na</span>(dif<span style="color:#f92672">$</span>EntrezID), <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">EntrezID&#34;</span>])

<span style="color:#75715e"># 设置GO富集分析的所有参数</span>
params <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">new</span>(<span style="color:#e6db74">&#34;</span><span style="color:#e6db74">GOHyperGParams&#34;</span>, geneIds<span style="color:#f92672">=</span>entrezSelected, universeGeneIds<span style="color:#f92672">=</span>entrezUniverse,
              annotation<span style="color:#f92672">=</span>affydb, ontology<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;</span><span style="color:#e6db74">BP&#34;</span>, pvalueCutoff<span style="color:#f92672">=</span><span style="color:#ae81ff">0.001</span>, conditional<span style="color:#f92672">=</span><span style="color:#66d9ef">FALSE</span>,
              testDirection<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;</span><span style="color:#e6db74">over&#34;</span>)
<span style="color:#75715e"># 对所有的GOterm根据params参数做超几何检验</span>
hgOver <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">hyperGTest</span>(params)

<span style="color:#75715e"># 生成所有GOterm的检验结果报表</span>
bp <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">summary</span>(hgOver)

<span style="color:#75715e"># 同时生成所有GOterm的检验结果文件，每个GOterm都有指向官方网站的链接，以获得详细信息</span>
<span style="color:#a6e22e">htmlReport</span>(hgOver, file<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;</span><span style="color:#e6db74">ALL_go.html&#34;</span>)

<span style="color:#75715e"># 显示结果前几行</span>
<span style="color:#a6e22e">head</span>(bp)
</code></pre></div><div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-R" data-lang="R"><span style="color:#f92672">&gt;</span> <span style="color:#a6e22e">head</span>(bp)
      GOBPID       Pvalue OddsRatio  ExpCount Count Size                                                     Term
<span style="color:#ae81ff">1</span> GO<span style="color:#f92672">:</span><span style="color:#ae81ff">0022904</span> <span style="color:#ae81ff">8.956477e-10</span>  <span style="color:#ae81ff">16.22041</span> <span style="color:#ae81ff">0.8868343</span>    <span style="color:#ae81ff">11</span>   <span style="color:#ae81ff">66</span>                     respiratory electron transport chain
<span style="color:#ae81ff">2</span> GO<span style="color:#f92672">:</span><span style="color:#ae81ff">0022900</span> <span style="color:#ae81ff">1.059675e-09</span>  <span style="color:#ae81ff">15.92875</span> <span style="color:#ae81ff">0.9002712</span>    <span style="color:#ae81ff">11</span>   <span style="color:#ae81ff">67</span>                                 electron transport chain
<span style="color:#ae81ff">3</span> GO<span style="color:#f92672">:</span><span style="color:#ae81ff">0042773</span> <span style="color:#ae81ff">3.220076e-09</span>  <span style="color:#ae81ff">17.09865</span> <span style="color:#ae81ff">0.7659024</span>    <span style="color:#ae81ff">10</span>   <span style="color:#ae81ff">57</span>                 ATP synthesis coupled electron transport
<span style="color:#ae81ff">4</span> GO<span style="color:#f92672">:</span><span style="color:#ae81ff">0042775</span> <span style="color:#ae81ff">3.220076e-09</span>  <span style="color:#ae81ff">17.09865</span> <span style="color:#ae81ff">0.7659024</span>    <span style="color:#ae81ff">10</span>   <span style="color:#ae81ff">57</span>   mitochondrial ATP synthesis coupled electron transport
<span style="color:#ae81ff">5</span> GO<span style="color:#f92672">:</span><span style="color:#ae81ff">0006123</span> <span style="color:#ae81ff">8.147710e-09</span>  <span style="color:#ae81ff">66.54092</span> <span style="color:#ae81ff">0.1746795</span>     <span style="color:#ae81ff">6</span>   <span style="color:#ae81ff">13</span> mitochondrial electron transport, cytochrome c to oxygen
<span style="color:#ae81ff">6</span> GO<span style="color:#f92672">:</span><span style="color:#ae81ff">0006119</span> <span style="color:#ae81ff">1.917217e-08</span>  <span style="color:#ae81ff">13.83664</span> <span style="color:#ae81ff">0.9137081</span>    <span style="color:#ae81ff">10</span>   <span style="color:#ae81ff">68</span>                                oxidative phosphorylation
</code></pre></div><div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-R" data-lang="R"><span style="color:#75715e"># 安装并加载所需包</span>
<span style="color:#a6e22e">biocLite</span>(<span style="color:#e6db74">&#34;</span><span style="color:#e6db74">GeneAnswers&#34;</span>)
<span style="color:#a6e22e">library</span>(GeneAnswers)

<span style="color:#75715e"># 选取dif中的三列信息构成新的矩阵，新一列必须是EntrezID</span>
humanGeneInput <span style="color:#f92672">&lt;-</span> dif[, <span style="color:#a6e22e">c</span>(<span style="color:#e6db74">&#34;</span><span style="color:#e6db74">EntrezID&#34;</span>, <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">logFC&#34;</span>, <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">P.Value&#34;</span>)]
<span style="color:#75715e"># 获得humanGeneInput中基因的表达值</span>
humanExpr <span style="color:#f92672">&lt;-</span> eset<span style="color:#a6e22e">[match</span>(<span style="color:#a6e22e">rownames</span>(dif), <span style="color:#a6e22e">rownames</span>(eset)), ]
humanExpr <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">cbind</span>(humanGeneInput[,<span style="color:#e6db74">&#34;</span><span style="color:#e6db74">EntrezID&#34;</span>], humanExpr)
<span style="color:#75715e"># 去除NA数据</span>
humanGeneInput <span style="color:#f92672">&lt;-</span> humanGeneInput[<span style="color:#f92672">!</span><span style="color:#a6e22e">is.na</span>(humanGeneInput[, <span style="color:#ae81ff">1</span>]),]
humanExpr <span style="color:#f92672">&lt;-</span> humanExpr[<span style="color:#f92672">!</span><span style="color:#a6e22e">is.na</span>(humanExpr[,<span style="color:#ae81ff">1</span>]),]
<span style="color:#75715e"># KEGG通路的超几何检验</span>
y <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">geneAnswersBuilder</span>(humanGeneInput, <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">org.Hs.eg.db&#34;</span>, categoryType <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">KEGG&#34;</span>,
                        testType <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">hyperG&#34;</span>, pvalueT<span style="color:#f92672">=</span><span style="color:#ae81ff">0.1</span>, geneExpressionProfile<span style="color:#f92672">=</span>humanExpr,
                        verbose <span style="color:#f92672">=</span> <span style="color:#66d9ef">FALSE</span>)
<span style="color:#a6e22e">getEnrichmentInfo</span>(y)[1<span style="color:#f92672">:</span><span style="color:#ae81ff">6</span>]
</code></pre></div><div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-R" data-lang="R"><span style="color:#f92672">&gt;</span> <span style="color:#a6e22e">getEnrichmentInfo</span>(y)[1<span style="color:#f92672">:</span><span style="color:#ae81ff">6</span>]
      genes in Category percent in the observed List percent in the genome fold of overrepresents odds ratio      p value
<span style="color:#ae81ff">05012</span>                <span style="color:#ae81ff">12</span>                   <span style="color:#ae81ff">0.22222222</span>           <span style="color:#ae81ff">0.022150281</span>              <span style="color:#ae81ff">10.032479</span>  <span style="color:#ae81ff">13.794189</span> <span style="color:#ae81ff">1.302401e-09</span>
<span style="color:#ae81ff">05010</span>                <span style="color:#ae81ff">13</span>                   <span style="color:#ae81ff">0.24074074</span>           <span style="color:#ae81ff">0.028454592</span>               <span style="color:#ae81ff">8.460523</span>  <span style="color:#ae81ff">11.655527</span> <span style="color:#ae81ff">2.022230e-09</span>
<span style="color:#ae81ff">00190</span>                <span style="color:#ae81ff">11</span>                   <span style="color:#ae81ff">0.20370370</span>           <span style="color:#ae81ff">0.022491055</span>               <span style="color:#ae81ff">9.057099</span>  <span style="color:#ae81ff">12.038055</span> <span style="color:#ae81ff">2.061276e-08</span>
<span style="color:#ae81ff">05016</span>                <span style="color:#ae81ff">12</span>                   <span style="color:#ae81ff">0.22222222</span>           <span style="color:#ae81ff">0.031180780</span>               <span style="color:#ae81ff">7.126897</span>   <span style="color:#ae81ff">9.430242</span> <span style="color:#ae81ff">6.436589e-08</span>
<span style="color:#ae81ff">04260</span>                 <span style="color:#ae81ff">7</span>                   <span style="color:#ae81ff">0.12962963</span>           <span style="color:#ae81ff">0.013119782</span>               <span style="color:#ae81ff">9.880471</span>  <span style="color:#ae81ff">12.223404</span> <span style="color:#ae81ff">5.483387e-06</span>
<span style="color:#ae81ff">01100</span>                <span style="color:#ae81ff">22</span>                   <span style="color:#ae81ff">0.40740741</span>           <span style="color:#ae81ff">0.192537059</span>               <span style="color:#ae81ff">2.115995</span>   <span style="color:#ae81ff">2.920634</span> <span style="color:#ae81ff">2.073075e-04</span>
<span style="color:#ae81ff">04710</span>                 <span style="color:#ae81ff">3</span>                   <span style="color:#ae81ff">0.05555556</span>           <span style="color:#ae81ff">0.003748509</span>              <span style="color:#ae81ff">14.820707</span>  <span style="color:#ae81ff">17.944272</span> <span style="color:#ae81ff">1.002054e-03</span>
<span style="color:#ae81ff">03040</span>                 <span style="color:#ae81ff">6</span>                   <span style="color:#ae81ff">0.11111111</span>           <span style="color:#ae81ff">0.021639121</span>               <span style="color:#ae81ff">5.134733</span>   <span style="color:#ae81ff">5.882231</span> <span style="color:#ae81ff">1.006287e-03</span>
<span style="color:#ae81ff">04310</span>                 <span style="color:#ae81ff">4</span>                   <span style="color:#ae81ff">0.07407407</span>           <span style="color:#ae81ff">0.025558017</span>               <span style="color:#ae81ff">2.898272</span>   <span style="color:#ae81ff">3.106301</span> <span style="color:#ae81ff">4.837838e-02</span>
<span style="color:#ae81ff">00230</span>                 <span style="color:#ae81ff">4</span>                   <span style="color:#ae81ff">0.07407407</span>           <span style="color:#ae81ff">0.027602658</span>               <span style="color:#ae81ff">2.683585</span>   <span style="color:#ae81ff">2.864304</span> <span style="color:#ae81ff">6.095794e-02</span>
<span style="color:#ae81ff">03050</span>                 <span style="color:#ae81ff">2</span>                   <span style="color:#ae81ff">0.03703704</span>           <span style="color:#ae81ff">0.007497018</span>               <span style="color:#ae81ff">4.940236</span>   <span style="color:#ae81ff">5.286630</span> <span style="color:#ae81ff">6.149150e-02</span>
<span style="color:#ae81ff">04114</span>                 <span style="color:#ae81ff">3</span>                   <span style="color:#ae81ff">0.05555556</span>           <span style="color:#ae81ff">0.019083319</span>               <span style="color:#ae81ff">2.911210</span>   <span style="color:#ae81ff">3.079331</span> <span style="color:#ae81ff">8.335034e-02</span>
<span style="color:#ae81ff">05221</span>                 <span style="color:#ae81ff">2</span>                   <span style="color:#ae81ff">0.03703704</span>           <span style="color:#ae81ff">0.009712046</span>               <span style="color:#ae81ff">3.813515</span>   <span style="color:#ae81ff">4.027972</span> <span style="color:#ae81ff">9.628046e-02</span>
</code></pre></div><p>上面调用GeneAnswers包实现了KEGG通路的注释、统计和可视化的功能。GeneAnswers功能强大，还支持GO、REACTOME和CABIO等多个数据库，可以通过设定参数categoryType分别指定注释类型。</p>
<p>根据已有的计算，我们接下来可视化分析结果，调用pheatmap包来绘制差异表达谱热图；调用Rgraphviz包会绘制显著富集的GO term的关系图；最后绘制显著富集的KEGG通路的关系图和热图。</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-R" data-lang="R"><span style="color:#a6e22e">biocLite</span>(<span style="color:#e6db74">&#34;</span><span style="color:#e6db74">pheatmap&#34;</span>)
<span style="color:#a6e22e">library</span>(pheatmap)
<span style="color:#75715e"># 从基因表达矩阵中，选取差异表达基因对应的数据</span>
selected <span style="color:#f92672">&lt;-</span> eset<span style="color:#a6e22e">[rownames</span>(dif), ]
<span style="color:#75715e"># 将selected矩阵每行的名称由探针组ID转换为对应的基因symbol</span>
<span style="color:#a6e22e">rownames</span>(selected) <span style="color:#f92672">&lt;-</span> dif<span style="color:#f92672">$</span>symbols
<span style="color:#75715e"># 考虑显示问题，这里只画前20个基因的热图</span>
<span style="color:#a6e22e">pheatmap</span>(selected[1<span style="color:#f92672">:</span><span style="color:#ae81ff">20</span>,], color <span style="color:#f92672">=</span> <span style="color:#a6e22e">colorRampPalette</span>(<span style="color:#a6e22e">c</span>(<span style="color:#e6db74">&#34;</span><span style="color:#e6db74">green&#34;</span>, <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">black&#34;</span>, <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">red&#34;</span>))(<span style="color:#ae81ff">100</span>),
         fontsize_row <span style="color:#f92672">=</span> <span style="color:#ae81ff">4</span>, scale <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">row&#34;</span>, border_color <span style="color:#f92672">=</span> <span style="color:#66d9ef">NA</span>)
</code></pre></div><p><strong>热图</strong></p>
<p><img src="../../../figures/pheatmap.png" alt=""></p>
<p>(截图原因，没有显示完全)</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-R" data-lang="R"><span style="color:#a6e22e">biocLite</span>(<span style="color:#e6db74">&#34;</span><span style="color:#e6db74">Rgraphviz&#34;</span>)
<span style="color:#a6e22e">library</span>(Rgraphviz)
<span style="color:#75715e"># 显著富集的GO term的DAG关系图</span>
ghandle <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">goDag</span>(hgOver)
<span style="color:#75715e"># 这个图太大，这里只能选一部分数据构建局部图</span>
subGHandle <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">subGraph</span>(snodes<span style="color:#f92672">=</span><span style="color:#a6e22e">as.character</span>(<span style="color:#a6e22e">summary</span>(hgOver)[,<span style="color:#ae81ff">1</span>]), graph <span style="color:#f92672">=</span> ghandle)
<span style="color:#a6e22e">plot</span>(subGHandle)
</code></pre></div><p><strong>显著富集的GO term关系图</strong></p>
<p><img src="../../../figures/subgraph_go.png" alt=""></p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-R" data-lang="R"><span style="color:#75715e"># 显著富集的KEGG通路的关系图</span>
yy <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">geneAnswersReadable</span>(y, verbose <span style="color:#f92672">=</span> <span style="color:#66d9ef">FALSE</span>)
<span style="color:#a6e22e">geneAnswersConceptNet</span>(yy, colorValueColumn <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">logFC&#34;</span>, centroidSize <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">pvalue&#34;</span>,
                      output <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">interactive&#34;</span>)
<span style="color:#75715e"># 显著富集的KEGG通路的热图</span>
yyy <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">geneAnswersSort</span>(yy, sortBy <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">pvalue&#34;</span>)
<span style="color:#a6e22e">geneAnswersHeatmap</span>(yyy)
</code></pre></div><p>（这里两个结果图因为我的依赖包配置有问题，还没有解决，大家可以自己试试。）</p>
<h2 id="heading-10">总结</h2>
<p>系统地走了一下基因芯片分析的流程，我们可以看到数据的分析主要分为两大块：预处理和分析。我们凭感觉往往侧重于分析，因为它们最靠近我们想要获得的结果。但从内容的介绍来看，我们可以观察到预处理部分富含更多的内容。分析的方法往往在短时间内会固化，近十年主流的分析过程主要的也就是提到的这几个，但是每个人想要处理的数据集是丰富多变的，它往往也决定了我们结果的可靠性。这是我们在分析时非常需要关注的一点。</p>
<p>最后输出会话信息：</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-R" data-lang="R"><span style="color:#f92672">&gt;</span> <span style="color:#a6e22e">sessionInfo</span>()
R version <span style="color:#ae81ff">3.4</span><span style="color:#ae81ff">.1</span> (<span style="color:#ae81ff">2017</span><span style="color:#ae81ff">-06</span><span style="color:#ae81ff">-30</span>)
Platform<span style="color:#f92672">:</span> x86_64<span style="color:#f92672">-</span>pc<span style="color:#f92672">-</span>linux<span style="color:#f92672">-</span><span style="color:#a6e22e">gnu </span>(<span style="color:#ae81ff">64</span><span style="color:#f92672">-</span>bit)
Running under<span style="color:#f92672">:</span> Ubuntu <span style="color:#ae81ff">17.04</span>

Matrix products<span style="color:#f92672">:</span> default
BLAS<span style="color:#f92672">:</span> <span style="color:#f92672">/</span>home<span style="color:#f92672">/</span>diviner<span style="color:#f92672">-</span>wsx<span style="color:#f92672">/</span>anaconda3<span style="color:#f92672">/</span>lib<span style="color:#f92672">/</span>R<span style="color:#f92672">/</span>lib<span style="color:#f92672">/</span>libRblas.so
LAPACK<span style="color:#f92672">:</span> <span style="color:#f92672">/</span>home<span style="color:#f92672">/</span>diviner<span style="color:#f92672">-</span>wsx<span style="color:#f92672">/</span>anaconda3<span style="color:#f92672">/</span>lib<span style="color:#f92672">/</span>R<span style="color:#f92672">/</span>lib<span style="color:#f92672">/</span>libRlapack.so

locale<span style="color:#f92672">:</span>
 [1] LC_CTYPE<span style="color:#f92672">=</span>en_US.UTF<span style="color:#ae81ff">-8</span>       LC_NUMERIC<span style="color:#f92672">=</span>C               LC_TIME<span style="color:#f92672">=</span>en_US.UTF<span style="color:#ae81ff">-8</span>        LC_COLLATE<span style="color:#f92672">=</span>en_US.UTF<span style="color:#ae81ff">-8</span>    
 [5] LC_MONETARY<span style="color:#f92672">=</span>en_US.UTF<span style="color:#ae81ff">-8</span>    LC_MESSAGES<span style="color:#f92672">=</span>en_US.UTF<span style="color:#ae81ff">-8</span>    LC_PAPER<span style="color:#f92672">=</span>en_US.UTF<span style="color:#ae81ff">-8</span>       LC_NAME<span style="color:#f92672">=</span>C                 
 [9] LC_ADDRESS<span style="color:#f92672">=</span>C               LC_TELEPHONE<span style="color:#f92672">=</span>C             LC_MEASUREMENT<span style="color:#f92672">=</span>en_US.UTF<span style="color:#ae81ff">-8</span> LC_IDENTIFICATION<span style="color:#f92672">=</span>C       

attached base packages<span style="color:#f92672">:</span>
 [1] grid      stats4    parallel  stats     graphics  grDevices utils     datasets  methods   base     

other attached packages<span style="color:#f92672">:</span>
 [1] Rgraphviz_2.20.0     GeneAnswers_2.18.0   RColorBrewer_1.1<span style="color:#ae81ff">-2</span>   Heatplus_2.22.0      MASS_7.3<span style="color:#ae81ff">-47</span>          RSQLite_2.0         
 [7] annotate_1.54.0      XML_3.98<span style="color:#ae81ff">-1.9</span>         RCurl_1.95<span style="color:#ae81ff">-4.8</span>       bitops_1.0<span style="color:#ae81ff">-6</span>         igraph_1.1.2         GOstats_2.42.0      
[13] Category_2.42.1      Matrix_1.2<span style="color:#ae81ff">-11</span>        AnnotationDbi_1.38.2 IRanges_2.10.4       S4Vectors_0.14.6     graph_1.54.0        
[19] limma_3.32.8         affy_1.54.0          Biobase_2.36.2       BiocGenerics_0.22.1 

loaded via a <span style="color:#a6e22e">namespace </span>(and not attached)<span style="color:#f92672">:</span>
 [1] Rcpp_0.12.13           BiocInstaller_1.26.1   compiler_3.4.1         tools_3.4.1            zlibbioc_1.22.0       
 [6] digest_0.6.12          bit_1.1<span style="color:#ae81ff">-12</span>             memoise_1.1.0          preprocessCore_1.38.1  tibble_1.3.4          
[11] lattice_0.20<span style="color:#ae81ff">-35</span>        pkgconfig_2.0.1        rlang_0.1.2            DBI_0.7                downloader_0.4        
[16] genefilter_1.58.1      bit64_0.9<span style="color:#ae81ff">-7</span>            GSEABase_1.38.2        RBGL_1.52.0            survival_2.41<span style="color:#ae81ff">-3</span>       
[21] magrittr_1.5           GO.db_3.4.1            blob_1.1.0             splines_3.4.1          AnnotationForge_1.18.2
[26] xtable_1.8<span style="color:#ae81ff">-2</span>           affyio_1.46.0         
</code></pre></div>

  <footer>
  
<nav class="post-nav">
  <span class="nav-prev">&larr; <a href="../../../cn/post/2017-09-04-control_shell/">控制脚本</a></span>
  <span class="nav-next"><a href="../../../cn/post/2017-10-27-rnaseq-data-analysis/">Bioconductor分析RNA-seq数据</a> &rarr;</span>
</nav>
<script type="text/javascript">
document.addEventListener('keyup', function(e) {
  if (e.target.nodeName.toUpperCase() != 'BODY') return;
  var url = false;
  if (e.which == 37) {  
    
    url = '\/cn\/post\/2017-09-04-control_shell\/';
    
  } else if (e.which == 39) {  
    
    url = '\/cn\/post\/2017-10-27-rnaseq-data-analysis\/';
    
  }
  if (url) window.location = url;
});
</script>



<section class="comments">
  <div id="disqus_thread"></div>
  <script src="../../../js/disqusloader.min.js"></script>
  <script>
  var disqus_config = function () {
  
    this.page.url = "https:\/\/shixiangwang.github.io\/home\/" + location.pathname;
  
  };
  (function() {
    var inIFrame = function() {
      var iframe = true;
      try { iframe = window.self !== window.top; } catch (e) {}
      return iframe;
    };
    if (inIFrame()) return;
    var disqus_js = '//home-xjdzylqrzp.disqus.com/embed.js';
    
    if (location.hash.match(/^#comment/)) {
      var d = document, s = d.createElement('script');
      s.src = disqus_js; s.async = true;
      s.setAttribute('data-timestamp', +new Date());
      (d.head || d.body).appendChild(s);
    } else {
      disqusLoader('#disqus_thread', {
        scriptUrl: disqus_js, laziness: 0, disqusConfig: disqus_config
      });
    }
  })();
  </script>
  <noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
</section>




<script async src="../../../js/fix-toc.js"></script>

<script async src="../../../js/center-img.js"></script>

<script async src="../../../js/right-quote.js"></script>

<script async src="../../../js/no-highlight.js"></script>

<script async src="../../../js/fix-footnote.js"></script>

<script async src="../../../js/math-code.js"></script>

<script async src="../../../js/external-link.js"></script>

<script async src="../../../js/alt-title.js"></script>

<script async src="../../../js/header-link.js"></script>


<script async src="//cdn.bootcss.com/mathjax/2.7.5/MathJax.js?config=TeX-MML-AM_CHTML"></script>

  



<script src="//cdn.bootcss.com/highlight.js/9.12.0/highlight.min.js"></script>



<script src="//cdn.bootcss.com/highlight.js/9.12.0/languages/r.min.js"></script>
<script src="//cdn.bootcss.com/highlight.js/9.12.0/languages/yaml.min.js"></script>
<script src="//cdn.bootcss.com/highlight.js/9.12.0/languages/tex.min.js"></script>
<script>hljs.configure({languages: []}); hljs.initHighlightingOnLoad();</script>




  
  
  

  <div class="copyright"><a href="mailto:w_shixiang@163.com"><!-- raw HTML omitted --><!-- raw HTML omitted --></a> · <a href="https://github.com/ShixiangWang"><!-- raw HTML omitted --><!-- raw HTML omitted --></a> · <a href="https://stackoverflow.com/users/7662327/shixiang-wang"><!-- raw HTML omitted --><!-- raw HTML omitted --></a> · <a href="https://scholar.google.com/citations?user=FvNp0NkAAAAJ&amp;hl=zh-CN"><!-- raw HTML omitted --><!-- raw HTML omitted --></a> · <a href="https://orcid.org/0000-0001-9855-7357"><!-- raw HTML omitted --><!-- raw HTML omitted --></a> · <a href="https://www.researchgate.net/profile/Wang_Shixiang4"><!-- raw HTML omitted --><!-- raw HTML omitted --></a> <!-- raw HTML omitted --> 本站由 <a href="https://gohugo.io">Hugo</a> 和 <a href="https://bookdown.org/yihui/blogdown/">Blogdown</a> 强力驱动 © <a href="../../../">王诗翔</a> 2017 - 2019 </div>
  
  

  <script type="text/javascript" src="//rf.revolvermaps.com/0/0/6.js?i=56h9es09xn7&amp;m=6&amp;c=ff0000&amp;cr1=ffffff&amp;f=arial&amp;l=1" async="async"></script>
  </footer>
  </article>
  
  </body>
</html>

